// tdesign-react/cjs/back-top/BackTop.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-5481698f.js');
var classNames = require('classnames');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useConfig = require('../hooks/useConfig.js');
var backTop_defaultProps = require('./defaultProps.js');
var backTop_useScroll = require('./useScroll.js');
var _util_dom = require('../_util/dom.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-dab07be2.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-abe9dd70.js');
require('../_chunks/dep-75a87f7c.js');
require('dayjs');
require('../_chunks/dep-f0496dae.js');
require('../_chunks/dep-89f1315e.js');
require('../_chunks/dep-49b692b4.js');
require('../_chunks/dep-5dc79997.js');
require('../_chunks/dep-350e0ec1.js');
require('../_chunks/dep-cef99009.js');
require('../_chunks/dep-a241cf5a.js');
require('../_chunks/dep-7a32c861.js');
require('../_chunks/dep-d4fd8ae6.js');
require('../_chunks/dep-de258415.js');
require('../_chunks/dep-b20ed9a4.js');
require('../_chunks/dep-ea572de0.js');
require('../_chunks/dep-ac56ab8f.js');
require('../_chunks/dep-e3f45bee.js');
require('../_chunks/dep-7ffb0d81.js');
require('../_chunks/dep-d02e27a7.js');
require('../_chunks/dep-dd8855d3.js');
require('../_chunks/dep-c2bb7040.js');
require('../_chunks/dep-ffef71d0.js');
require('../_chunks/dep-53b02ac8.js');
require('../_chunks/dep-54d84720.js');
require('../_chunks/dep-da766de6.js');
require('../_chunks/dep-b20af383.js');
require('../_chunks/dep-7225af62.js');
require('raf');
require('../_util/easing.js');
require('../_chunks/dep-b2af1b16.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getContainer = function getContainer(container) {
  if (typeof container === "string") {
    if (typeof document !== "undefined") {
      return document.querySelector(container);
    }
  }
  if (typeof container === "function") {
    return container();
  }
  return null;
};
var InternalBackTop = function InternalBackTop(props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, backTop_defaultProps.backTopDefaultProps),
    theme = _useDefaultProps.theme,
    size = _useDefaultProps.size,
    shape = _useDefaultProps.shape,
    target = _useDefaultProps.target,
    visibleHeight = _useDefaultProps.visibleHeight,
    container = _useDefaultProps.container,
    duration = _useDefaultProps.duration,
    content = _useDefaultProps.content,
    offset = _useDefaultProps.offset,
    children = _useDefaultProps.children,
    cusContent = _useDefaultProps["default"],
    className = _useDefaultProps.className,
    style = _useDefaultProps.style,
    onClick = _useDefaultProps.onClick;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var scrollContainer = React.useMemo(function () {
    return getContainer(container);
  }, [container]);
  var _useScroll = backTop_useScroll["default"]({
      target: scrollContainer
    }),
    scrollTop = _useScroll.scrollTop;
  var defaultContent = /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.BacktopIcon, {
    className: "".concat(classPrefix, "-back-top__icon"),
    size: 24
  }), /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-back-top__text")
  }, "TOP"));
  var renderChildren = children || content || cusContent || defaultContent;
  var backTopRef = React__default["default"].useRef(null);
  React__default["default"].useImperativeHandle(ref, function () {
    return backTopRef.current;
  });
  var visible = React.useMemo(function () {
    if (typeof visibleHeight === "string") {
      return scrollTop >= Number(visibleHeight.replace("px", ""));
    }
    return scrollTop >= visibleHeight;
  }, [scrollTop, visibleHeight]);
  var backTopStyle = React.useMemo(function () {
    return _objectSpread({
      insetInlineEnd: offset[0],
      insetBlockEnd: offset[1]
    }, style);
  }, [offset, style]);
  var cls = classNames__default["default"]("".concat(classPrefix, "-back-top"), "".concat(classPrefix, "-back-top--theme-").concat(theme), "".concat(classPrefix, "-back-top--").concat(shape), defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-back-top--show"), visible), "".concat(classPrefix, "-size-s"), size === "small"), "".concat(classPrefix, "-size-m"), size === "medium"), className);
  var getBackTo = React.useCallback(function () {
    if (!target) return 0;
    var targetNode = getContainer(target);
    if (!targetNode) return 0;
    var rect = targetNode.getBoundingClientRect();
    var y = rect.y;
    return y;
  }, [target]);
  var handleClick = React.useCallback(function (e) {
    e.stopPropagation();
    onClick === null || onClick === void 0 || onClick({
      e: e
    });
    var backTo = getBackTo();
    _util_dom.scrollTo(backTo, {
      container: scrollContainer,
      duration: duration
    });
  }, [duration, getBackTo, onClick, scrollContainer]);
  return /* @__PURE__ */React__default["default"].createElement("button", {
    type: "button",
    ref: backTopRef,
    className: cls,
    style: backTopStyle,
    onClick: handleClick
  }, renderChildren);
};
var BackTop = /*#__PURE__*/React__default["default"].forwardRef(InternalBackTop);
BackTop.displayName = "BackTop";

exports["default"] = BackTop;
