// tdesign-react/cjs/auto-complete/AutoComplete.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-5481698f.js');
var slicedToArray = require('../_chunks/dep-b20af383.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useControlled = require('../hooks/useControlled.js');
var autoComplete_defaultProps = require('./defaultProps.js');
var hooks_useCommonClassName = require('../hooks/useCommonClassName.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var input_index = require('../input/index.js');
var popup_index = require('../popup/index.js');
var autoComplete_OptionList = require('./OptionList.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-dab07be2.js');
require('../_chunks/dep-7225af62.js');
require('../_util/noop.js');
require('../_chunks/dep-2a4a8f70.js');
require('../_chunks/dep-ba13d061.js');
require('../_chunks/dep-1e0bc473.js');
require('../_chunks/dep-fff57180.js');
require('../_chunks/dep-5dc79997.js');
require('../_chunks/dep-cef99009.js');
require('../_chunks/dep-ac9a26ed.js');
require('../_chunks/dep-350e0ec1.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-abe9dd70.js');
require('../_chunks/dep-75a87f7c.js');
require('dayjs');
require('../_chunks/dep-f0496dae.js');
require('../_chunks/dep-89f1315e.js');
require('../_chunks/dep-49b692b4.js');
require('../_chunks/dep-a241cf5a.js');
require('../_chunks/dep-7a32c861.js');
require('../_chunks/dep-d4fd8ae6.js');
require('../_chunks/dep-de258415.js');
require('../_chunks/dep-b20ed9a4.js');
require('../_chunks/dep-ea572de0.js');
require('../_chunks/dep-ac56ab8f.js');
require('../_chunks/dep-e3f45bee.js');
require('../_chunks/dep-7ffb0d81.js');
require('../_chunks/dep-d02e27a7.js');
require('../_chunks/dep-dd8855d3.js');
require('../_chunks/dep-c2bb7040.js');
require('../_chunks/dep-ffef71d0.js');
require('../_chunks/dep-53b02ac8.js');
require('../_chunks/dep-54d84720.js');
require('../_chunks/dep-da766de6.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-040a2db4.js');
require('../_chunks/dep-641538cb.js');
require('../_chunks/dep-b686b3fb.js');
require('../_chunks/dep-307d0db8.js');
require('../_chunks/dep-b3648f6a.js');
require('../_chunks/dep-62fd65d7.js');
require('../_chunks/dep-62978241.js');
require('../input/Input.js');
require('../_chunks/dep-d19a672c.js');
require('tdesign-icons-react');
require('../hooks/useLayoutEffect.js');
require('../_util/dom.js');
require('raf');
require('../_util/easing.js');
require('../_chunks/dep-b2af1b16.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../hooks/useGlobalIcon.js');
require('../input/InputGroup.js');
require('../input/defaultProps.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-e831f387.js');
require('../input/useLengthLimit.js');
require('../_chunks/dep-be7119b7.js');
require('../_chunks/dep-b691f3a6.js');
require('../_chunks/dep-ca323c2b.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../hooks/useAnimation.js');
require('../common/Portal.js');
require('react-dom');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../_util/ref.js');
require('../_util/isFragment.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-3e4c9d3d.js');
require('../_chunks/dep-9bee38bb.js');
require('../_chunks/dep-0ba932e9.js');
require('../_chunks/dep-a8e37273.js');
require('../_chunks/dep-ae011087.js');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../hooks/useAttach.js');
require('../hooks/usePopper.js');
require('@popperjs/core');
require('react-fast-compare');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-2b39f9ea.js');
require('../_util/react-render.js');
require('react-dom/client');
require('../common/PluginContainer.js');
require('../_chunks/dep-ae281384.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var AutoComplete = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var _props$placeholder;
  var props = hooks_useDefaultProps["default"](originalProps, autoComplete_defaultProps.autoCompleteDefaultProps);
  var inputRef = React.useRef(null);
  var popupRef = React.useRef(null);
  var _useControlled = hooks_useControlled["default"](props, "value", props.onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    tValue = _useControlled2[0],
    setTValue = _useControlled2[1];
  var _useCommonClassName = hooks_useCommonClassName["default"](),
    classPrefix = _useCommonClassName.classPrefix,
    sizeClassNames = _useCommonClassName.sizeClassNames;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("input"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 1),
    global = _useLocaleReceiver2[0];
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    popupVisible = _useState2[0],
    setPopupVisible = _useState2[1];
  var optionListRef = React.useRef(null);
  React.useImperativeHandle(ref, function () {
    return {
      inputRef: inputRef.current,
      popupRef: popupRef.current
    };
  });
  var getOverlayStyle = function getOverlayStyle(trigger, popupElement) {
    var _props$popupProps;
    var triggerWidth = trigger.getBoundingClientRect().width || trigger.offsetWidth || trigger.clientWidth;
    var popupWidth = popupElement.getBoundingClientRect().width || popupElement.offsetWidth || popupElement.clientWidth;
    return _objectSpread({
      width: triggerWidth >= popupWidth ? "".concat(triggerWidth, "px") : "auto"
    }, (_props$popupProps = props.popupProps) === null || _props$popupProps === void 0 ? void 0 : _props$popupProps.overlayInnerStyle);
  };
  var classes = ["".concat(classPrefix, "-auto-complete")].concat(props.className);
  var popupClasses = function (_props$popupProps2) {
    var classes2 = ["".concat(classPrefix, "-select__dropdown")];
    if ((_props$popupProps2 = props.popupProps) !== null && _props$popupProps2 !== void 0 && _props$popupProps2.overlayClassName) {
      classes2 = classes2.concat(props.popupProps.overlayClassName);
    }
    return classNames__default["default"](classes2);
  }();
  var popupInnerClasses = function (_props$popupProps3) {
    var classes2 = ["".concat(classPrefix, "-select__dropdown-inner")];
    if ((_props$popupProps3 = props.popupProps) !== null && _props$popupProps3 !== void 0 && _props$popupProps3.overlayInnerClassName) {
      classes2 = classes2.concat(props.popupProps.overlayInnerClassName);
    }
    return classNames__default["default"](classes2);
  }();
  var onInputChange = function onInputChange(value, context) {
    if (!popupVisible) {
      setPopupVisible(true);
    }
    setTValue(value, context);
  };
  var innerInputProps = function () {
    var tProps = _objectSpread({
      value: tValue,
      size: props.size
    }, props.inputProps);
    return tProps;
  }();
  var onInnerFocus = function onInnerFocus(value, context) {
    var _props$onFocus;
    setPopupVisible(true);
    (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 || _props$onFocus.call(props, _objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
    var timer = setTimeout(function () {
      var _optionListRef$curren;
      (_optionListRef$curren = optionListRef.current) === null || _optionListRef$curren === void 0 || _optionListRef$curren.addKeyboardListener();
      clearTimeout(timer);
    }, 0);
  };
  var onInnerBlur = function onInnerBlur(value, context) {
    var _props$onBlur;
    (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 || _props$onBlur.call(props, _objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
  };
  var onInnerEnter = function onInnerEnter(value, context) {
    var _props$onEnter;
    (_props$onEnter = props.onEnter) === null || _props$onEnter === void 0 || _props$onEnter.call(props, _objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
  };
  var onInnerCompositionend = function onInnerCompositionend(value, context) {
    var _props$onCompositione;
    (_props$onCompositione = props.onCompositionend) === null || _props$onCompositione === void 0 || _props$onCompositione.call(props, _objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
  };
  var onInnerCompositionstart = function onInnerCompositionstart(value, context) {
    var _props$onCompositions;
    (_props$onCompositions = props.onCompositionstart) === null || _props$onCompositions === void 0 || _props$onCompositions.call(props, _objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
  };
  var onInnerSelect = function onInnerSelect(value, context) {
    var _props$onSelect;
    if (props.readonly || props.disabled) return;
    setPopupVisible(false);
    setTValue(value, context);
    (_props$onSelect = props.onSelect) === null || _props$onSelect === void 0 || _props$onSelect.call(props, value, context);
  };
  var onPopupVisibleChange = function onPopupVisibleChange(visible, _ref) {
    var trigger = _ref.trigger;
    if (trigger !== "trigger-element-click") {
      setPopupVisible(visible);
    }
  };
  var triggerNode = props.triggerElement || props.children || /* @__PURE__ */React__default["default"].createElement(input_index.Input, _objectSpread({
    ref: inputRef,
    placeholder: (_props$placeholder = props.placeholder) !== null && _props$placeholder !== void 0 ? _props$placeholder : global.placeholder,
    tips: props.tips,
    status: props.status,
    readonly: props.readonly,
    disabled: props.disabled,
    clearable: props.clearable,
    autofocus: props.autofocus,
    borderless: props.borderless,
    onClear: props.onClear,
    onChange: onInputChange,
    onFocus: onInnerFocus,
    onBlur: onInnerBlur,
    onEnter: onInnerEnter,
    onCompositionend: onInnerCompositionend,
    onCompositionstart: onInnerCompositionstart
  }, innerInputProps));
  var listContent = Array.isArray(props.options) && /* @__PURE__ */React__default["default"].createElement(autoComplete_OptionList["default"], {
    ref: optionListRef,
    value: tValue,
    options: props.options,
    size: props.size,
    sizeClassNames: sizeClassNames,
    onSelect: onInnerSelect,
    popupVisible: popupVisible,
    highlightKeyword: props.highlightKeyword,
    filterable: props.filterable,
    filter: props.filter,
    empty: props.empty
  });
  var topContent = props.panelTopContent;
  var bottomContent = props.panelBottomContent;
  var panelContent = topContent || listContent || bottomContent ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-auto-complete__panel")
  }, topContent, listContent, bottomContent) : null;
  var popupProps = _objectSpread(_objectSpread({}, props.popupProps), {}, {
    overlayInnerStyle: getOverlayStyle,
    overlayInnerClassName: popupInnerClasses,
    overlayClassName: popupClasses
  });
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](classes),
    style: props.style
  }, /* @__PURE__ */React__default["default"].createElement(popup_index.Popup, _objectSpread({
    ref: popupRef,
    visible: popupVisible,
    onVisibleChange: onPopupVisibleChange,
    trigger: "focus",
    placement: "bottom-left",
    hideEmptyPopup: true,
    content: panelContent
  }, popupProps), triggerNode));
});
AutoComplete.displayName = "AutoComplete";

exports["default"] = AutoComplete;
