// tdesign-react/cjs/select/hooks/useOptions.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-5481698f.js');
var slicedToArray = require('../../_chunks/dep-b20af383.js');
var React = require('react');
var select_util_helper = require('../util/helper.js');
var select_base_Option = require('../base/Option.js');
var select_base_OptionGroup = require('../base/OptionGroup.js');
var get = require('../../_chunks/dep-e831607b.js');
require('../../_chunks/dep-dab07be2.js');
require('../../_chunks/dep-7225af62.js');
require('../../_chunks/dep-b691f3a6.js');
require('../../_chunks/dep-54d84720.js');
require('../../_chunks/dep-5dc79997.js');
require('../../_chunks/dep-c2bb7040.js');
require('../../_chunks/dep-d02e27a7.js');
require('../../_chunks/dep-dd8855d3.js');
require('../../_chunks/dep-e3f45bee.js');
require('../../_chunks/dep-d4fd8ae6.js');
require('../../_chunks/dep-de258415.js');
require('../../_chunks/dep-7ffb0d81.js');
require('../../_chunks/dep-89f1315e.js');
require('../../_chunks/dep-49b692b4.js');
require('../../_chunks/dep-350e0ec1.js');
require('../../_chunks/dep-cef99009.js');
require('../../_chunks/dep-a241cf5a.js');
require('../../_chunks/dep-7a32c861.js');
require('../../_chunks/dep-b20ed9a4.js');
require('../../_chunks/dep-ea572de0.js');
require('../../_chunks/dep-ac56ab8f.js');
require('classnames');
require('../../hooks/useConfig.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-abe9dd70.js');
require('../../_chunks/dep-75a87f7c.js');
require('dayjs');
require('../../_chunks/dep-f0496dae.js');
require('../../_chunks/dep-ffef71d0.js');
require('../../_chunks/dep-53b02ac8.js');
require('../../_chunks/dep-da766de6.js');
require('../defaultProps.js');
require('../../hooks/useDefaultProps.js');
require('../../hooks/useDomRefCallback.js');
require('../../hooks/useRipple.js');
require('../../hooks/useAnimation.js');
require('../../_util/setStyle.js');
require('../../_util/dom.js');
require('raf');
require('../../_util/easing.js');
require('../../_chunks/dep-b2af1b16.js');
require('../../_chunks/dep-ca323c2b.js');
require('../../_chunks/dep-29a8a75a.js');
require('../../_chunks/dep-ac9a26ed.js');
require('../../_chunks/dep-fff57180.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function UseOptions(keys, options, children, valueType, value, reserveKeyword) {
  var _useState = React.useState({}),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    valueToOption = _useState2[0],
    setValueToOption = _useState2[1];
  var _useState3 = React.useState([]),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    currentOptions = _useState4[0],
    setCurrentOptions = _useState4[1];
  var _useState5 = React.useState([]),
    _useState6 = slicedToArray._slicedToArray(_useState5, 2),
    tmpPropOptions = _useState6[0],
    setTmpPropOptions = _useState6[1];
  var _useState7 = React.useState([]),
    _useState8 = slicedToArray._slicedToArray(_useState7, 2),
    selectedOptions = _useState8[0],
    setSelectedOptions = _useState8[1];
  React.useEffect(function () {
    var transformedOptions = options;
    var arrayChildren = React__default["default"].Children.toArray(children);
    var optionChildren = arrayChildren.filter(function (v) {
      return v.type === select_base_Option["default"] || v.type === select_base_OptionGroup["default"];
    });
    var isChildrenFilterable = arrayChildren.length > 0 && optionChildren.length === arrayChildren.length;
    if (reserveKeyword && currentOptions.length && isChildrenFilterable) return;
    if (isChildrenFilterable) {
      var _handlerOptionElement = function handlerOptionElement(v) {
        if (/*#__PURE__*/React__default["default"].isValidElement(v)) {
          if (v.type === select_base_OptionGroup["default"]) {
            var _v$props$children;
            return _objectSpread(_objectSpread({}, v.props), {}, {
              group: v.props.label,
              children: (_v$props$children = v.props.children) === null || _v$props$children === void 0 ? void 0 : _v$props$children.map(function (v2) {
                return _handlerOptionElement(v2);
              })
            });
          }
          return _objectSpread(_objectSpread({}, v.props), {}, {
            label: v.props.label || v.props.children
          });
        }
        return {
          label: v
        };
      };
      transformedOptions = arrayChildren === null || arrayChildren === void 0 ? void 0 : arrayChildren.map(function (v) {
        return _handlerOptionElement(v);
      });
    }
    if (keys) {
      var _transformedOptions;
      transformedOptions = (_transformedOptions = transformedOptions) === null || _transformedOptions === void 0 ? void 0 : _transformedOptions.map(function (option) {
        return _objectSpread(_objectSpread({}, option), {}, {
          value: get.get(option, (keys === null || keys === void 0 ? void 0 : keys.value) || "value"),
          label: get.get(option, (keys === null || keys === void 0 ? void 0 : keys.label) || "label")
        });
      });
    }
    setCurrentOptions(transformedOptions);
    setTmpPropOptions(transformedOptions);
    setValueToOption(select_util_helper.getValueToOption(children, options, keys) || {});
  }, [options, keys, children, reserveKeyword]);
  React.useEffect(function () {
    setSelectedOptions(function (oldSelectedOptions) {
      var valueKey = (keys === null || keys === void 0 ? void 0 : keys.value) || "value";
      var labelKey = (keys === null || keys === void 0 ? void 0 : keys.label) || "label";
      if (Array.isArray(value)) {
        return value.map(function (item) {
          if (valueType === "value") {
            return valueToOption[item] || oldSelectedOptions.find(function (option) {
              return get.get(option, valueKey) === item;
            }) || defineProperty._defineProperty(defineProperty._defineProperty({}, valueKey, item), labelKey, item);
          }
          return item;
        }).filter(Boolean);
      }
      if (value !== void 0 && value !== null) {
        if (valueType === "value") {
          return [valueToOption[value] || oldSelectedOptions.find(function (option) {
            return get.get(option, valueKey) === value;
          }) || defineProperty._defineProperty(defineProperty._defineProperty({}, valueKey, value), labelKey, value)].filter(Boolean);
        }
        return [value];
      }
      return [];
    });
  }, [value, keys, valueType, valueToOption, setSelectedOptions]);
  return {
    currentOptions: currentOptions,
    setCurrentOptions: setCurrentOptions,
    tmpPropOptions: tmpPropOptions,
    setTmpPropOptions: setTmpPropOptions,
    valueToOption: valueToOption,
    setValueToOption: setValueToOption,
    selectedOptions: selectedOptions,
    setSelectedOptions: setSelectedOptions
  };
}

exports["default"] = UseOptions;
