// tdesign-react/cjs/form/FormContext.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var FormContext = /*#__PURE__*/React__default["default"].createContext({
  form: void 0,
  labelWidth: "100px",
  labelAlign: "right",
  layout: "vertical",
  colon: false,
  initialData: {},
  requiredMark: void 0,
  scrollToFirstError: void 0,
  showErrorMessage: void 0,
  resetType: "empty",
  disabled: void 0,
  rules: void 0,
  errorMessage: void 0,
  statusIcon: void 0,
  onFormItemValueChange: void 0,
  formMapRef: void 0,
  floatingFormDataRef: void 0
});
var useFormContext = function useFormContext() {
  return React__default["default"].useContext(FormContext);
};
var FormListContext = /*#__PURE__*/React__default["default"].createContext({
  name: void 0,
  rules: void 0,
  formListMapRef: void 0,
  initialData: []
});
var useFormListContext = function useFormListContext() {
  return React__default["default"].useContext(FormListContext);
};

exports.FormListContext = FormListContext;
exports["default"] = FormContext;
exports.useFormContext = useFormContext;
exports.useFormListContext = useFormListContext;
