// tdesign-react/lib/date-picker/panel/SinglePanel.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-daac9b4e.js';
import React, { forwardRef } from 'react';
import classNames from 'classnames';
import { isFunction } from 'lodash-es';
import { g as getDefaultFormat, p as parseToDayjs } from '../../_chunks/dep-81a8909d.js';
import useConfig from '../../hooks/useConfig.js';
import PanelContent from './PanelContent.js';
import ExtraContent from './ExtraContent.js';
import useTableData from '../hooks/useTableData.js';
import useDisableDate from '../hooks/useDisableDate.js';
import useDefaultProps from '../../hooks/useDefaultProps.js';
import { parseToDateTime } from '../utils.js';
import '../../_chunks/dep-2b61bedd.js';
import 'dayjs';
import '../../_chunks/dep-d1906e4e.js';
import '../../_chunks/dep-abe679d5.js';
import '../../config-provider/ConfigContext.js';
import '../../_chunks/dep-d2abf305.js';
import '../base/Header.js';
import '../../_chunks/dep-385f8c20.js';
import '../../_chunks/dep-73b0b578.js';
import '../../_chunks/dep-17424654.js';
import '../../locale/LocalReceiver.js';
import '../../config-provider/ConfigProvider.js';
import '../../select/index.js';
import '../../select/base/Select.js';
import '../../_chunks/dep-b209b5b7.js';
import '../../_util/helper.js';
import '../../hooks/useControlled.js';
import '../../_util/noop.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../select/util/helper.js';
import '../../select/base/OptionGroup.js';
import '../../select/defaultProps.js';
import '../../select/base/Option.js';
import '../../hooks/useDomRefCallback.js';
import '../../hooks/useRipple.js';
import '../../hooks/useAnimation.js';
import '../../_util/setStyle.js';
import '../../_util/dom.js';
import 'raf';
import '../../_util/easing.js';
import '../../common/FakeArrow.js';
import '../../loading/index.js';
import '../../loading/Loading.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../hooks/useLayoutEffect.js';
import '../../_chunks/dep-5ac87a7c.js';
import '../../_chunks/dep-f7978f59.js';
import '../../loading/defaultProps.js';
import '../../loading/plugin.js';
import '../../_util/react-render.js';
import '../../_chunks/dep-2c5bf7ef.js';
import 'react-dom/client';
import '../../common/PluginContainer.js';
import '../../select-input/index.js';
import '../../select-input/SelectInput.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import '../../popup/hooks/useTrigger.js';
import 'react-is';
import '../../_util/ref.js';
import '../../_util/isFragment.js';
import '../../_util/composeRefs.js';
import '../../popup/utils/transition.js';
import '../../hooks/useMutationObserver.js';
import '../../hooks/useLatest.js';
import '../../hooks/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../hooks/useAttach.js';
import '../../hooks/usePopper.js';
import '@popperjs/core';
import 'react-fast-compare';
import '../../popup/PopupPlugin.js';
import '../../select-input/useSingle.js';
import '../../input/index.js';
import '../../input/Input.js';
import 'tdesign-icons-react';
import '../../hooks/useGlobalIcon.js';
import '../../input/InputGroup.js';
import '../../input/defaultProps.js';
import '../../_util/parseTNode.js';
import '../../input/useLengthLimit.js';
import '../../select-input/useMultiple.js';
import '../../tag-input/index.js';
import '../../tag-input/TagInput.js';
import '../../hooks/useDragSorter.js';
import '../../hooks/useEventCallback.js';
import '../../tag-input/useTagScroll.js';
import '../../tag-input/useTagList.js';
import '../../tag/index.js';
import '../../tag/Tag.js';
import 'tinycolor2';
import '../../tag/defaultProps.js';
import '../../tag/CheckTag.js';
import '../../_chunks/dep-e91d193e.js';
import '../../tag/CheckTagGroup.js';
import '../../tag-input/useHover.js';
import '../../tag-input/defaultProps.js';
import '../../select-input/useOverlayInnerStyle.js';
import '../../select-input/defaultProps.js';
import '../../select/base/PopupContent.js';
import '../../select/hooks/usePanelVirtualScroll.js';
import '../../hooks/useVirtualScroll.js';
import '../../hooks/useResizeObserver.js';
import '../../select/hooks/useOptions.js';
import '../../pagination/index.js';
import '../../pagination/Pagination.js';
import '../../input-number/index.js';
import '../../input-number/InputNumber.js';
import '../../button/index.js';
import '../../button/Button.js';
import '../../button/defaultProps.js';
import '../../input-number/useInputNumber.js';
import '../../_chunks/dep-5a2b6e16.js';
import '../../hooks/useCommonClassName.js';
import '../../input-number/defaultProps.js';
import '../../input-adornment/InputAdornment.js';
import '../../pagination/hooks/useBoundaryJumper.js';
import '../../pagination/hooks/usePrevNextJumper.js';
import '../../pagination/hooks/usePageNumber.js';
import '../../pagination/hooks/useTotal.js';
import '../../pagination/validators.js';
import '../../pagination/defaultProps.js';
import '../../pagination/PaginationMini.js';
import '../base/Table.js';
import '../base/Cell.js';
import '../../_chunks/dep-f5a6f5d9.js';
import '../../_chunks/dep-9ec15c19.js';
import '../../time-picker/index.js';
import '../../time-picker/TimePicker.js';
import '../../_chunks/dep-f7dd0aec.js';
import '../../time-picker/hooks/useTimePickerTextConfig.js';
import '../../hooks/useDebounce.js';
import '../../hooks/usePersistFn.js';
import '../../hooks/usePropsRef.js';
import '../../time-picker/TimeRangePicker.js';
import '../../range-input/index.js';
import '../../range-input/RangeInput.js';
import '../../range-input/defaultProps.js';
import '../../range-input/RangeInputPopup.js';
import '../../time-picker/panel/TimePickerPanel.js';
import '../../time-picker/defaultProps.js';
import '../base/Footer.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var SinglePanel = /*#__PURE__*/forwardRef(function (originalProps, ref) {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix,
    globalDatePickerConfig = _useConfig.datePicker;
  var panelName = "".concat(classPrefix, "-date-picker__panel");
  var props = useDefaultProps(originalProps, {
    mode: "date",
    enableTimePicker: false,
    presetsPlacement: "bottom"
  });
  var value = props.value,
    mode = props.mode,
    presetsPlacement = props.presetsPlacement,
    _props$firstDayOfWeek = props.firstDayOfWeek,
    firstDayOfWeek = _props$firstDayOfWeek === void 0 ? globalDatePickerConfig.firstDayOfWeek : _props$firstDayOfWeek,
    needConfirm = props.needConfirm,
    style = props.style,
    className = props.className,
    year = props.year,
    month = props.month,
    onPanelClick = props.onPanelClick,
    disableTime = props.disableTime;
  var _getDefaultFormat = getDefaultFormat({
      mode: props.mode,
      format: props.format,
      enableTimePicker: props.enableTimePicker
    }),
    format = _getDefaultFormat.format;
  var disableDateOptions = useDisableDate({
    disableDate: props.disableDate,
    mode: props.mode,
    format: format
  });
  var disableTimeOptions = function disableTimeOptions(h, m, s, ms) {
    if (!isFunction(disableTime) || !value) {
      return {};
    }
    return disableTime(parseToDateTime(value, format, [h, m, s, ms]));
  };
  var tableData = useTableData(_objectSpread({
    value: value,
    year: year,
    month: month,
    mode: mode,
    start: value ? parseToDayjs(props.multiple ? value[0] : value, format).toDate() : void 0,
    firstDayOfWeek: firstDayOfWeek,
    multiple: props.multiple
  }, disableDateOptions));
  var panelContentProps = {
    mode: mode,
    value: value,
    year: year,
    month: month,
    format: format,
    firstDayOfWeek: firstDayOfWeek,
    tableData: tableData,
    popupVisible: props.popupVisible,
    multiple: props.multiple,
    time: props.time,
    timePickerProps: _objectSpread({
      disableTime: disableTimeOptions
    }, props.timePickerProps),
    enableTimePicker: props.enableTimePicker,
    onMonthChange: props.onMonthChange,
    onYearChange: props.onYearChange,
    onJumperClick: props.onJumperClick,
    onCellClick: props.onCellClick,
    onCellMouseEnter: props.onCellMouseEnter,
    onCellMouseLeave: props.onCellMouseLeave,
    onTimePickerChange: props.onTimePickerChange
  };
  var extraProps = {
    presets: props.presets,
    enableTimePicker: props.enableTimePicker,
    presetsPlacement: props.presetsPlacement,
    onPresetClick: props.onPresetClick,
    onConfirmClick: props.onConfirmClick,
    selectedValue: props.value,
    needConfirm: needConfirm
  };
  return /* @__PURE__ */React.createElement("div", {
    ref: ref,
    style: style,
    className: classNames(panelName, className, _defineProperty({}, "".concat(panelName, "--direction-row"), ["left", "right"].includes(presetsPlacement))),
    onClick: function onClick(e) {
      return onPanelClick === null || onPanelClick === void 0 ? void 0 : onPanelClick({
        e: e
      });
    }
  }, ["top", "left"].includes(presetsPlacement) ? /* @__PURE__ */React.createElement(ExtraContent, _objectSpread({}, extraProps)) : null, /* @__PURE__ */React.createElement(PanelContent, _objectSpread({}, panelContentProps)), ["bottom", "right"].includes(presetsPlacement) ? /* @__PURE__ */React.createElement(ExtraContent, _objectSpread({}, extraProps)) : null);
});
SinglePanel.displayName = "SinglePanel";

export { SinglePanel as default };
