// tdesign-react/cjs/pagination/PaginationMini.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-5481698f.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var button_index = require('../button/index.js');
var pagination_defaultProps = require('./defaultProps.js');
var _util_noop = require('../_util/noop.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var isObject = require('../_chunks/dep-de258415.js');
require('../_chunks/dep-dab07be2.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-abe9dd70.js');
require('../_chunks/dep-75a87f7c.js');
require('dayjs');
require('../_chunks/dep-f0496dae.js');
require('../_chunks/dep-89f1315e.js');
require('../_chunks/dep-49b692b4.js');
require('../_chunks/dep-5dc79997.js');
require('../_chunks/dep-350e0ec1.js');
require('../_chunks/dep-cef99009.js');
require('../_chunks/dep-a241cf5a.js');
require('../_chunks/dep-7a32c861.js');
require('../_chunks/dep-d4fd8ae6.js');
require('../_chunks/dep-b20ed9a4.js');
require('../_chunks/dep-ea572de0.js');
require('../_chunks/dep-ac56ab8f.js');
require('../_chunks/dep-e3f45bee.js');
require('../_chunks/dep-7ffb0d81.js');
require('../_chunks/dep-d02e27a7.js');
require('../_chunks/dep-dd8855d3.js');
require('../_chunks/dep-c2bb7040.js');
require('../_chunks/dep-ffef71d0.js');
require('../_chunks/dep-53b02ac8.js');
require('../_chunks/dep-54d84720.js');
require('../_chunks/dep-da766de6.js');
require('../button/Button.js');
require('../_chunks/dep-b20af383.js');
require('../_chunks/dep-7225af62.js');
require('../_chunks/dep-d19a672c.js');
require('../hooks/useDomRefCallback.js');
require('../hooks/useRipple.js');
require('../hooks/useAnimation.js');
require('../_util/setStyle.js');
require('../_util/dom.js');
require('raf');
require('../_util/easing.js');
require('../_chunks/dep-b2af1b16.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../_chunks/dep-05ad44c3.js');
require('../_chunks/dep-be7119b7.js');
require('../_chunks/dep-b691f3a6.js');
require('../_chunks/dep-ca323c2b.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-2b39f9ea.js');
require('react-dom/client');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-040a2db4.js');
require('../_chunks/dep-641538cb.js');
require('../_chunks/dep-b686b3fb.js');
require('../_chunks/dep-307d0db8.js');
require('../_chunks/dep-b3648f6a.js');
require('../_chunks/dep-62fd65d7.js');
require('../_chunks/dep-62978241.js');
require('../button/defaultProps.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-e831f387.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var PaginationMini = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      RoundIcon: tdesignIconsReact.RoundIcon,
      ChevronUpIcon: tdesignIconsReact.ChevronUpIcon,
      ChevronDownIcon: tdesignIconsReact.ChevronDownIcon,
      ChevronLeftIcon: tdesignIconsReact.ChevronLeftIcon,
      ChevronRightIcon: tdesignIconsReact.ChevronRightIcon
    }),
    RoundIcon = _useGlobalIcon.RoundIcon,
    ChevronUpIcon = _useGlobalIcon.ChevronUpIcon,
    ChevronDownIcon = _useGlobalIcon.ChevronDownIcon,
    ChevronLeftIcon = _useGlobalIcon.ChevronLeftIcon,
    ChevronRightIcon = _useGlobalIcon.ChevronRightIcon;
  var _useDefaultProps = hooks_useDefaultProps["default"](props, pagination_defaultProps.paginationMiniDefaultProps),
    variant = _useDefaultProps.variant,
    tips = _useDefaultProps.tips,
    showCurrent = _useDefaultProps.showCurrent,
    disabled = _useDefaultProps.disabled,
    layout = _useDefaultProps.layout,
    size = _useDefaultProps.size,
    _useDefaultProps$onCh = _useDefaultProps.onChange,
    onChange = _useDefaultProps$onCh === void 0 ? _util_noop["default"] : _useDefaultProps$onCh,
    className = _useDefaultProps.className,
    style = _useDefaultProps.style;
  var titleConfig = React.useMemo(function () {
    if (isObject.isObject(tips)) {
      return tips;
    }
    if (tips === true) {
      return {
        prev: "\u4E0A\u4E00\u9875",
        current: "\u5F53\u524D",
        next: "\u4E0B\u4E00\u9875"
      };
    }
    return {};
  }, [tips]);
  var disabledConfig = React.useMemo(function () {
    if (isObject.isObject(disabled)) {
      return disabled;
    }
    if (disabled === true) {
      return {
        prev: true,
        current: true,
        next: true
      };
    }
    return {
      prev: false,
      current: false,
      next: false
    };
  }, [disabled]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-pagination-mini"), className, defineProperty._defineProperty({}, "".concat(classPrefix, "-pagination-mini--outline"), variant === "outline")),
    ref: ref,
    style: style
  }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    title: titleConfig.prev,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "prev"
      });
    },
    icon: layout === "horizontal" ? /* @__PURE__ */React__default["default"].createElement(ChevronLeftIcon, null) : /* @__PURE__ */React__default["default"].createElement(ChevronUpIcon, null),
    className: "".concat(classPrefix, "-pagination-mini__prev"),
    disabled: disabledConfig.prev
  }), showCurrent && /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    title: titleConfig.current,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "current"
      });
    },
    icon: /* @__PURE__ */React__default["default"].createElement(RoundIcon, null),
    className: "".concat(classPrefix, "-pagination-mini__current"),
    disabled: disabledConfig.current
  }), /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    title: titleConfig.next,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "next"
      });
    },
    icon: layout === "horizontal" ? /* @__PURE__ */React__default["default"].createElement(ChevronRightIcon, null) : /* @__PURE__ */React__default["default"].createElement(ChevronDownIcon, null),
    className: "".concat(classPrefix, "-pagination-mini__next"),
    disabled: disabledConfig.next
  }));
});
PaginationMini.displayName = "PaginationMini";

exports["default"] = PaginationMini;
