// tdesign-react/cjs/tree/hooks/useDraggable.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-5481698f.js');
var slicedToArray = require('../../_chunks/dep-b20af383.js');
var React = require('react');
var tree_hooks_TreeDraggableContext = require('./TreeDraggableContext.js');
var hooks_usePersistFn = require('../../hooks/usePersistFn.js');
var throttle = require('../../_chunks/dep-02e4e6c8.js');
require('../../_chunks/dep-dab07be2.js');
require('../../_chunks/dep-7225af62.js');
require('../../_util/createHookContext.js');
require('../../_chunks/dep-a8e37273.js');
require('../../_chunks/dep-de258415.js');
require('../../_chunks/dep-5dc79997.js');
require('../../_chunks/dep-ae011087.js');
require('../../_chunks/dep-ac9a26ed.js');
require('../../_chunks/dep-350e0ec1.js');

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function useDraggable(props) {
  var nodeRef = props.nodeRef,
    node = props.node,
    allowDrop = props.allowDrop;
  var _useTreeDraggableCont = tree_hooks_TreeDraggableContext.useTreeDraggableContext(),
    onDragStart = _useTreeDraggableCont.onDragStart,
    onDragEnd = _useTreeDraggableCont.onDragEnd,
    onDragLeave = _useTreeDraggableCont.onDragLeave,
    onDragOver = _useTreeDraggableCont.onDragOver,
    onDrop = _useTreeDraggableCont.onDrop;
  var _useState = React.useState({
      isDragOver: false,
      isDragging: false,
      dropPosition: 0
    }),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    state = _useState2[0],
    setState = _useState2[1];
  var setPartialState = hooks_usePersistFn.usePersistFn(function (newState) {
    setState(function (prev) {
      return _objectSpread(_objectSpread({}, prev), newState);
    });
  });
  var updateDropPosition = React.useRef(throttle.throttle(function (e) {
    if (!nodeRef.current) return;
    var rect = nodeRef.current.getBoundingClientRect();
    var offsetY = window.pageYOffset + rect.top;
    var pageY = e.pageY;
    var gapHeight = rect.height / 4;
    var diff = pageY - offsetY;
    if (diff < gapHeight) {
      setPartialState({
        dropPosition: -1
      });
    } else if (diff < rect.height - gapHeight) {
      setPartialState({
        dropPosition: 0
      });
    } else {
      setPartialState({
        dropPosition: 1
      });
    }
  })).current;
  var setDragStatus = function setDragStatus(status, e) {
    switch (status) {
      case "dragStart":
        setPartialState({
          isDragging: true,
          dropPosition: 0
        });
        onDragStart === null || onDragStart === void 0 || onDragStart({
          node: node,
          e: e
        });
        break;
      case "dragEnd":
        setPartialState({
          isDragging: false,
          isDragOver: false,
          dropPosition: 0
        });
        updateDropPosition.cancel();
        onDragEnd === null || onDragEnd === void 0 || onDragEnd({
          node: node,
          e: e
        });
        break;
      case "dragOver":
        setPartialState({
          isDragOver: true
        });
        updateDropPosition(e);
        onDragOver === null || onDragOver === void 0 || onDragOver({
          node: node,
          e: e
        });
        break;
      case "dragLeave":
        setPartialState({
          isDragOver: false,
          dropPosition: 0
        });
        updateDropPosition.cancel();
        onDragLeave === null || onDragLeave === void 0 || onDragLeave({
          node: node,
          e: e
        });
        break;
      case "drop":
        onDrop === null || onDrop === void 0 || onDrop({
          node: node,
          dropPosition: state.dropPosition,
          e: e,
          allowDrop: allowDrop
        });
        setPartialState({
          isDragOver: false
        });
        updateDropPosition.cancel();
        break;
    }
  };
  return _objectSpread(_objectSpread({}, state), {}, {
    setDragStatus: setDragStatus
  });
}

exports["default"] = useDraggable;
