// tdesign-react/cjs/date-picker/DateRangePickerPanel.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-5481698f.js');
var toConsumableArray = require('../_chunks/dep-b691f3a6.js');
var slicedToArray = require('../_chunks/dep-b20af383.js');
var React = require('react');
var dayjs = require('dayjs');
var format = require('../_chunks/dep-5b1d510b.js');
var utils = require('../_chunks/dep-cb44acad.js');
var log = require('../_chunks/dep-e831f387.js');
var datePicker_panel_RangePanel = require('./panel/RangePanel.js');
var datePicker_hooks_useRangeValue = require('./hooks/useRangeValue.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var datePicker_utils = require('./utils.js');
require('../_chunks/dep-dab07be2.js');
require('../_chunks/dep-7225af62.js');
require('../_chunks/dep-75a87f7c.js');
require('../_chunks/dep-b2af1b16.js');
require('../_chunks/dep-5dc79997.js');
require('../_chunks/dep-cef99009.js');
require('../_chunks/dep-350e0ec1.js');
require('../_chunks/dep-4de03639.js');
require('../_chunks/dep-ae011087.js');
require('../_chunks/dep-de258415.js');
require('../_chunks/dep-ac9a26ed.js');
require('../_chunks/dep-d4fd8ae6.js');
require('../_chunks/dep-1e0bc473.js');
require('../_chunks/dep-da766de6.js');
require('../_chunks/dep-7ffb0d81.js');
require('../_chunks/dep-7a32c861.js');
require('../_chunks/dep-ea572de0.js');
require('classnames');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-abe9dd70.js');
require('../_chunks/dep-f0496dae.js');
require('../_chunks/dep-89f1315e.js');
require('../_chunks/dep-49b692b4.js');
require('../_chunks/dep-a241cf5a.js');
require('../_chunks/dep-b20ed9a4.js');
require('../_chunks/dep-ac56ab8f.js');
require('../_chunks/dep-e3f45bee.js');
require('../_chunks/dep-d02e27a7.js');
require('../_chunks/dep-dd8855d3.js');
require('../_chunks/dep-c2bb7040.js');
require('../_chunks/dep-ffef71d0.js');
require('../_chunks/dep-53b02ac8.js');
require('../_chunks/dep-54d84720.js');
require('./panel/PanelContent.js');
require('./base/Header.js');
require('../locale/LocalReceiver.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-040a2db4.js');
require('../_chunks/dep-641538cb.js');
require('../_chunks/dep-b686b3fb.js');
require('../_chunks/dep-307d0db8.js');
require('../_chunks/dep-b3648f6a.js');
require('../_chunks/dep-62fd65d7.js');
require('../_chunks/dep-62978241.js');
require('../select/index.js');
require('../select/base/Select.js');
require('../_chunks/dep-d19a672c.js');
require('../_util/helper.js');
require('../_chunks/dep-10ac59e4.js');
require('../_chunks/dep-fff57180.js');
require('../_chunks/dep-2a4a8f70.js');
require('../_chunks/dep-ba13d061.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../select/util/helper.js');
require('../select/base/OptionGroup.js');
require('../select/defaultProps.js');
require('../select/base/Option.js');
require('../hooks/useDomRefCallback.js');
require('../hooks/useRipple.js');
require('../hooks/useAnimation.js');
require('../_util/setStyle.js');
require('../_util/dom.js');
require('raf');
require('../_util/easing.js');
require('../_chunks/dep-ca323c2b.js');
require('../_chunks/dep-e831607b.js');
require('../_chunks/dep-29a8a75a.js');
require('../common/FakeArrow.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../_chunks/dep-05ad44c3.js');
require('../_chunks/dep-be7119b7.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-2b39f9ea.js');
require('react-dom/client');
require('../common/PluginContainer.js');
require('../select-input/index.js');
require('../select-input/SelectInput.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../_util/ref.js');
require('../_util/isFragment.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-3e4c9d3d.js');
require('../_chunks/dep-9bee38bb.js');
require('../_chunks/dep-0ba932e9.js');
require('../_chunks/dep-a8e37273.js');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../hooks/useAttach.js');
require('../hooks/usePopper.js');
require('@popperjs/core');
require('react-fast-compare');
require('../popup/PopupPlugin.js');
require('../select-input/useSingle.js');
require('../input/index.js');
require('../input/Input.js');
require('tdesign-icons-react');
require('../hooks/useGlobalIcon.js');
require('../input/InputGroup.js');
require('../input/defaultProps.js');
require('../_util/parseTNode.js');
require('../input/useLengthLimit.js');
require('../_chunks/dep-3c90fe9e.js');
require('../_chunks/dep-3e385230.js');
require('../_chunks/dep-079c57d9.js');
require('../_chunks/dep-c74cf9fd.js');
require('../_chunks/dep-781be47c.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('../hooks/useDragSorter.js');
require('../hooks/useEventCallback.js');
require('../tag-input/useTagScroll.js');
require('../tag-input/useTagList.js');
require('../tag/index.js');
require('../tag/Tag.js');
require('tinycolor2');
require('../tag/defaultProps.js');
require('../tag/CheckTag.js');
require('../_chunks/dep-9cc6ea5f.js');
require('../tag/CheckTagGroup.js');
require('../tag-input/useHover.js');
require('../tag-input/defaultProps.js');
require('../select-input/useOverlayInnerStyle.js');
require('../select-input/defaultProps.js');
require('../select/base/PopupContent.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../hooks/useVirtualScroll.js');
require('../hooks/useResizeObserver.js');
require('../select/hooks/useOptions.js');
require('../pagination/index.js');
require('../pagination/Pagination.js');
require('../input-number/index.js');
require('../input-number/InputNumber.js');
require('../button/index.js');
require('../button/Button.js');
require('../button/defaultProps.js');
require('../input-number/useInputNumber.js');
require('../_chunks/dep-68a3b90a.js');
require('../_chunks/dep-0f33907c.js');
require('../hooks/useCommonClassName.js');
require('../input-number/defaultProps.js');
require('../input-adornment/InputAdornment.js');
require('../pagination/hooks/useBoundaryJumper.js');
require('../pagination/hooks/usePrevNextJumper.js');
require('../pagination/hooks/usePageNumber.js');
require('../pagination/hooks/useTotal.js');
require('../pagination/validators.js');
require('../pagination/defaultProps.js');
require('../_chunks/dep-be9ef47b.js');
require('../_chunks/dep-8b483a12.js');
require('../pagination/PaginationMini.js');
require('./base/Table.js');
require('./base/Cell.js');
require('../time-picker/index.js');
require('../time-picker/TimePicker.js');
require('../_chunks/dep-c217409c.js');
require('../time-picker/hooks/useTimePickerTextConfig.js');
require('../hooks/useDebounce.js');
require('../hooks/usePersistFn.js');
require('../hooks/usePropsRef.js');
require('../time-picker/TimeRangePicker.js');
require('../range-input/index.js');
require('../range-input/RangeInput.js');
require('../range-input/defaultProps.js');
require('../range-input/RangeInputPopup.js');
require('../time-picker/panel/TimePickerPanel.js');
require('../time-picker/defaultProps.js');
require('./panel/ExtraContent.js');
require('./base/Footer.js');
require('./hooks/useTableData.js');
require('./hooks/useDisableDate.js');
require('../_chunks/dep-d554fd33.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var DateRangePickerPanel = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var props = hooks_useDefaultProps["default"](originalProps, {
    mode: "date",
    defaultValue: [],
    panelPreselection: true
  });
  var className = props.className,
    style = props.style,
    mode = props.mode,
    enableTimePicker = props.enableTimePicker,
    disableDate = props.disableDate,
    firstDayOfWeek = props.firstDayOfWeek,
    presets = props.presets,
    timePickerProps = props.timePickerProps,
    presetsPlacement = props.presetsPlacement,
    panelPreselection = props.panelPreselection,
    onPanelClick = props.onPanelClick;
  var _useRangeValue = datePicker_hooks_useRangeValue["default"](props),
    value = _useRangeValue.value,
    year = _useRangeValue.year,
    month = _useRangeValue.month,
    _useRangeValue$time = _useRangeValue.time,
    time = _useRangeValue$time === void 0 ? [] : _useRangeValue$time,
    onChange = _useRangeValue.onChange,
    setTime = _useRangeValue.setTime,
    setYear = _useRangeValue.setYear,
    setMonth = _useRangeValue.setMonth,
    isFirstValueSelected = _useRangeValue.isFirstValueSelected,
    setIsFirstValueSelected = _useRangeValue.setIsFirstValueSelected,
    cacheValue = _useRangeValue.cacheValue,
    setCacheValue = _useRangeValue.setCacheValue;
  var _getDefaultFormat = format.getDefaultFormat({
      mode: mode,
      enableTimePicker: enableTimePicker,
      format: props.format
    }),
    format$1 = _getDefaultFormat.format;
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isSelected = _useState2[0],
    setIsSelected = _useState2[1];
  var _useState3 = React.useState(false),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    isHoverCell = _useState4[0],
    setIsHoverCell = _useState4[1];
  var _useState5 = React.useState([]),
    _useState6 = slicedToArray._slicedToArray(_useState5, 2),
    hoverValue = _useState6[0],
    setHoverValue = _useState6[1];
  var activeIndex = React.useMemo(function () {
    return isFirstValueSelected ? 1 : 0;
  }, [isFirstValueSelected]);
  function onCellMouseEnter(date) {
    setIsHoverCell(true);
    var nextValue = toConsumableArray._toConsumableArray(hoverValue);
    nextValue[activeIndex] = format.formatDate(date, {
      format: format$1
    });
    setHoverValue(nextValue);
  }
  function onCellMouseLeave() {
    setIsHoverCell(false);
    setHoverValue(cacheValue);
  }
  function onCellClick(date, _ref) {
    var _props$onCellClick;
    var e = _ref.e,
      partial = _ref.partial;
    setIsSelected(true);
    var nextValue = toConsumableArray._toConsumableArray(cacheValue);
    nextValue[activeIndex] = format.formatDate(date, {
      format: format$1
    });
    setCacheValue(nextValue);
    (_props$onCellClick = props.onCellClick) === null || _props$onCellClick === void 0 || _props$onCellClick.call(props, {
      date: nextValue.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      e: e,
      partial: activeIndex ? "end" : "start"
    });
    if (mode === "date") {
      var partialIndex = partial === "start" ? 0 : 1;
      var isAdditional = dayjs__default["default"](date).month() !== month[partialIndex];
      if (isAdditional) {
        if (activeIndex === 0) setMonth([dayjs__default["default"](date).month(), Math.min(dayjs__default["default"](date).month() + 1, 11)]);
        if (activeIndex === 1) setMonth([Math.max(dayjs__default["default"](date).month() - 1, 0), dayjs__default["default"](date).month()]);
      }
    }
    if (enableTimePicker) return;
    if (nextValue.length === 2 && isFirstValueSelected) {
      onChange(format.formatDate(nextValue, {
        format: format$1,
        autoSwap: true
      }), {
        dayjsValue: nextValue.map(function (v) {
          return format.parseToDayjs(v, format$1);
        }),
        trigger: "pick"
      });
      setIsFirstValueSelected(false);
    } else {
      setIsFirstValueSelected(true);
    }
  }
  function onJumperClick(_ref2) {
    var trigger = _ref2.trigger,
      partial = _ref2.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    var triggerMap = {
      "-1": "arrow-previous",
      1: "arrow-next"
    };
    var monthCountMap = {
      date: 1,
      week: 1,
      month: 12,
      quarter: 12,
      year: 120
    };
    var monthCount = monthCountMap[mode] || 0;
    var current = new Date(year[partialIndex], month[partialIndex]);
    var next = null;
    if (trigger === "prev") {
      next = utils.subtractMonth(current, monthCount);
    } else if (trigger === "current") {
      next = new Date();
    } else if (trigger === "next") {
      next = utils.addMonth(current, monthCount);
    }
    var nextYear = toConsumableArray._toConsumableArray(year);
    nextYear[partialIndex] = next.getFullYear();
    var nextMonth = toConsumableArray._toConsumableArray(month);
    nextMonth[partialIndex] = next.getMonth();
    var onlyYearSelect = ["year", "quarter", "month"].includes(mode);
    var correctedDate = datePicker_utils.dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
    nextYear = correctedDate.nextYear;
    nextMonth = correctedDate.nextMonth;
    if (year.some(function (y) {
      return !nextYear.includes(y);
    })) {
      var _props$onYearChange;
      (_props$onYearChange = props.onYearChange) === null || _props$onYearChange === void 0 || _props$onYearChange.call(props, {
        partial: partial,
        year: nextYear[partialIndex],
        date: value.map(function (v) {
          return format.parseToDayjs(v, format$1).toDate();
        }),
        trigger: trigger === "current" ? "today" : "year-".concat(triggerMap[trigger])
      });
    }
    if (month.some(function (m) {
      return !nextMonth.includes(m);
    })) {
      var _props$onMonthChange;
      (_props$onMonthChange = props.onMonthChange) === null || _props$onMonthChange === void 0 || _props$onMonthChange.call(props, {
        partial: partial,
        month: nextMonth[partialIndex],
        date: value.map(function (v) {
          return format.parseToDayjs(v, format$1).toDate();
        }),
        trigger: trigger === "current" ? "today" : "month-".concat(triggerMap[trigger])
      });
    }
    setYear(nextYear);
    setMonth(nextMonth);
  }
  function onTimePickerChange(val) {
    var _props$onTimeChange;
    var _extractTimeObj = utils.extractTimeObj(val),
      hours = _extractTimeObj.hours,
      minutes = _extractTimeObj.minutes,
      seconds = _extractTimeObj.seconds,
      milliseconds = _extractTimeObj.milliseconds,
      meridiem = _extractTimeObj.meridiem;
    var nextInputValue = toConsumableArray._toConsumableArray(cacheValue);
    var changedInputValue = cacheValue[activeIndex];
    var currentDate = !dayjs__default["default"](changedInputValue, format$1).isValid() ? dayjs__default["default"]().year(year[activeIndex]).month(month[activeIndex]) : dayjs__default["default"](changedInputValue, format$1);
    var nextHours = hours;
    if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
    if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
    var nextDate = currentDate.hour(nextHours).minute(minutes).second(seconds).millisecond(milliseconds).toDate();
    nextInputValue[activeIndex] = nextDate;
    var nextTime = toConsumableArray._toConsumableArray(time);
    nextTime[activeIndex] = val;
    setTime(nextTime);
    setIsSelected(true);
    setCacheValue(format.formatDate(nextInputValue, {
      format: format$1
    }));
    (_props$onTimeChange = props.onTimeChange) === null || _props$onTimeChange === void 0 || _props$onTimeChange.call(props, {
      time: val,
      partial: activeIndex ? "end" : "start",
      date: value.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      trigger: "time-hour"
    });
  }
  function onConfirmClick(_ref3) {
    var _props$onConfirm;
    var e = _ref3.e;
    var nextValue = toConsumableArray._toConsumableArray(cacheValue);
    if (nextValue.length === 2 && isFirstValueSelected) {
      onChange(format.formatDate(nextValue, {
        format: format$1,
        autoSwap: true
      }), {
        dayjsValue: nextValue.map(function (v) {
          return format.parseToDayjs(v, format$1);
        }),
        trigger: "confirm"
      });
      setYear(nextValue.map(function (v) {
        return dayjs__default["default"](v, format$1).year();
      }));
      setMonth(nextValue.map(function (v) {
        return dayjs__default["default"](v, format$1).month();
      }));
      setIsFirstValueSelected(false);
    } else {
      setIsFirstValueSelected(true);
    }
    (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 || _props$onConfirm.call(props, {
      date: value.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      e: e
    });
  }
  function onPresetClick(presetValue, context) {
    var _props$onPresetClick;
    var presetVal = typeof presetValue === "function" ? presetValue() : presetValue;
    if (!Array.isArray(presetVal)) {
      log.log.error("DateRangePickerPanel", "preset: ".concat(presetValue, " must be Array!"));
    } else {
      onChange(format.formatDate(presetVal, {
        format: format$1,
        autoSwap: true
      }), {
        dayjsValue: presetVal.map(function (p) {
          return format.parseToDayjs(p, format$1);
        }),
        trigger: "preset"
      });
    }
    (_props$onPresetClick = props.onPresetClick) === null || _props$onPresetClick === void 0 || _props$onPresetClick.call(props, context);
  }
  function onYearChange(nextVal, _ref4) {
    var _props$onYearChange2;
    var partial = _ref4.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    if (enableTimePicker) partialIndex = activeIndex;
    var nextYear = toConsumableArray._toConsumableArray(year);
    nextYear[partialIndex] = nextVal;
    var nextMonth = toConsumableArray._toConsumableArray(month);
    nextYear[partialIndex] = nextVal;
    var onlyYearSelect = ["year", "quarter", "month"].includes(mode);
    var correctedDate = datePicker_utils.dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
    nextYear = correctedDate.nextYear;
    nextMonth = correctedDate.nextMonth;
    setYear(nextYear);
    !onlyYearSelect && setMonth(nextMonth);
    (_props$onYearChange2 = props.onYearChange) === null || _props$onYearChange2 === void 0 || _props$onYearChange2.call(props, {
      partial: partial,
      year: nextYear[partialIndex],
      date: value.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      trigger: "year-select"
    });
  }
  function onMonthChange(nextVal, _ref5) {
    var _props$onMonthChange2;
    var partial = _ref5.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    if (enableTimePicker) partialIndex = activeIndex;
    var nextMonth = toConsumableArray._toConsumableArray(month);
    nextMonth[partialIndex] = nextVal;
    if (year[0] === year[1]) {
      if (partialIndex === 0) {
        if (nextMonth[1] <= nextMonth[0]) {
          nextMonth[1] = nextMonth[0] + 1;
          if (nextMonth[1] === 12) {
            nextMonth[1] = 0;
            setYear(function (currentYear) {
              return [currentYear[0], currentYear[1] + 1];
            });
          }
        }
      }
      if (partialIndex === 1) {
        nextMonth[0] = Math.min(nextMonth[0], nextMonth[1]);
        if (nextMonth[0] >= nextMonth[1]) {
          nextMonth[0] -= 1;
          if (nextMonth[0] === -1) {
            nextMonth[0] = 11;
            setYear(function (currentYear) {
              return [currentYear[0] - 1, currentYear[1]];
            });
          }
        }
      }
    }
    setMonth(nextMonth);
    (_props$onMonthChange2 = props.onMonthChange) === null || _props$onMonthChange2 === void 0 || _props$onMonthChange2.call(props, {
      partial: partial,
      month: nextMonth[partialIndex],
      date: value.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      trigger: "month-select"
    });
  }
  var panelProps = {
    hoverValue: isHoverCell ? hoverValue : [],
    value: isSelected ? cacheValue : value,
    year: year,
    month: month,
    mode: mode,
    format: format$1,
    presets: presets,
    time: time,
    disableDate: disableDate,
    firstDayOfWeek: firstDayOfWeek,
    timePickerProps: timePickerProps,
    enableTimePicker: enableTimePicker,
    presetsPlacement: presetsPlacement,
    activeIndex: activeIndex,
    panelPreselection: panelPreselection,
    onCellMouseEnter: onCellMouseEnter,
    onCellMouseLeave: onCellMouseLeave,
    onCellClick: onCellClick,
    onJumperClick: onJumperClick,
    onConfirmClick: onConfirmClick,
    onPresetClick: onPresetClick,
    onYearChange: onYearChange,
    onMonthChange: onMonthChange,
    onTimePickerChange: onTimePickerChange,
    onPanelClick: onPanelClick
  };
  return /* @__PURE__ */React__default["default"].createElement(datePicker_panel_RangePanel["default"], _objectSpread({
    ref: ref,
    className: className,
    style: style
  }, panelProps));
});
DateRangePickerPanel.displayName = "DateRangePickerPanel";

exports["default"] = DateRangePickerPanel;
