// tdesign-react/es/empty/assets/MaintenanceSvg.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

import React from 'react';

function MaintenanceSvg() {
  return /* @__PURE__ */React.createElement("svg", {
    width: "1em",
    height: "1em",
    viewBox: "0 0 48 48",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /* @__PURE__ */React.createElement("g", {
    id: "\xE8\xB7\xAF\xE5\x86\xB5\xE4\xBA\x8B\xE4\xBB\xB6-traffic-events"
  }, /* @__PURE__ */React.createElement("g", {
    id: "Subtract"
  }, /* @__PURE__ */React.createElement("path", {
    d: "M29.5237 17L24 3.82812L18.4763 17H29.5237Z",
    fill: "var(--td-text-color-placeholder)"
  }), /* @__PURE__ */React.createElement("path", {
    d: "M31.2011 21H16.7989L13.6699 28.4615H34.3301L31.2011 21Z",
    fill: "var(--td-text-color-placeholder)"
  }), /* @__PURE__ */React.createElement("path", {
    d: "M36.0076 32.4615H11.9924L9.66997 37.9997H6V41.9997H42V37.9997H38.33L36.0076 32.4615Z",
    fill: "var(--td-text-color-placeholder)"
  }))));
}

export { MaintenanceSvg as default };
