// tdesign-react/cjs/table/Cell.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _typeof = require('../_chunks/dep-dab07be2.js');
var defineProperty = require('../_chunks/dep-5481698f.js');
var React = require('react');
var classNames = require('classnames');
var table_Ellipsis = require('./Ellipsis.js');
var table_hooks_useFixed = require('./hooks/useFixed.js');
var table_utils = require('./utils.js');
var isFunction = require('../_chunks/dep-d4fd8ae6.js');
var get = require('../_chunks/dep-e831607b.js');
require('../_chunks/dep-b20af383.js');
require('../_chunks/dep-7225af62.js');
require('../_util/dom.js');
require('raf');
require('../_util/easing.js');
require('../_chunks/dep-b2af1b16.js');
require('../_chunks/dep-5dc79997.js');
require('../_chunks/dep-cef99009.js');
require('../_chunks/dep-350e0ec1.js');
require('../tooltip/index.js');
require('../tooltip/Tooltip.js');
require('../_chunks/dep-d19a672c.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-2a4a8f70.js');
require('../_chunks/dep-ba13d061.js');
require('../_chunks/dep-1e0bc473.js');
require('../_chunks/dep-fff57180.js');
require('../_chunks/dep-ac9a26ed.js');
require('../hooks/useAnimation.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-abe9dd70.js');
require('../_chunks/dep-75a87f7c.js');
require('dayjs');
require('../_chunks/dep-f0496dae.js');
require('../_chunks/dep-89f1315e.js');
require('../_chunks/dep-49b692b4.js');
require('../_chunks/dep-a241cf5a.js');
require('../_chunks/dep-7a32c861.js');
require('../_chunks/dep-b20ed9a4.js');
require('../_chunks/dep-ea572de0.js');
require('../_chunks/dep-ac56ab8f.js');
require('../_chunks/dep-e3f45bee.js');
require('../_chunks/dep-de258415.js');
require('../_chunks/dep-7ffb0d81.js');
require('../_chunks/dep-d02e27a7.js');
require('../_chunks/dep-dd8855d3.js');
require('../_chunks/dep-c2bb7040.js');
require('../_chunks/dep-ffef71d0.js');
require('../_chunks/dep-53b02ac8.js');
require('../_chunks/dep-54d84720.js');
require('../_chunks/dep-da766de6.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../_util/ref.js');
require('../_util/isFragment.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-3e4c9d3d.js');
require('../_chunks/dep-9bee38bb.js');
require('../_chunks/dep-0ba932e9.js');
require('../_chunks/dep-b686b3fb.js');
require('../_chunks/dep-307d0db8.js');
require('../_chunks/dep-b3648f6a.js');
require('../_chunks/dep-62fd65d7.js');
require('../_chunks/dep-62978241.js');
require('../_chunks/dep-a8e37273.js');
require('../_chunks/dep-ae011087.js');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../hooks/useDefaultProps.js');
require('../hooks/useAttach.js');
require('../hooks/usePopper.js');
require('../_chunks/dep-b691f3a6.js');
require('@popperjs/core');
require('react-fast-compare');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-2b39f9ea.js');
require('../_util/react-render.js');
require('react-dom/client');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-040a2db4.js');
require('../_chunks/dep-641538cb.js');
require('../tooltip/defaultProps.js');
require('../tooltip/TooltipLite.js');
require('../hooks/useSwitch.js');
require('../hooks/usePersistFn.js');
require('../_chunks/dep-02e4e6c8.js');
require('../hooks/useDebounce.js');
require('../_chunks/dep-be7119b7.js');
require('../_chunks/dep-ca323c2b.js');
require('../_chunks/dep-e831f387.js');
require('../_chunks/dep-b9a2cb3a.js');
require('../hooks/usePrevious.js');
require('../_chunks/dep-0ab4d869.js');
require('../_chunks/dep-cdb82b24.js');
require('../_chunks/dep-90a6d62f.js');
require('../_chunks/dep-781be47c.js');
require('../_chunks/dep-3ca282ce.js');
require('../_chunks/dep-8b483a12.js');
require('../_chunks/dep-3c90fe9e.js');
require('../_chunks/dep-29a8a75a.js');
require('../_chunks/dep-3e385230.js');
require('../_chunks/dep-079c57d9.js');
require('../_chunks/dep-c74cf9fd.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function renderCell(params, extra) {
  var col = params.col,
    row = params.row,
    rowIndex = params.rowIndex;
  if (col.colKey === "serial-number") {
    var _ref = (extra === null || extra === void 0 ? void 0 : extra.pagination) || {},
      current = _ref.current,
      pageSize = _ref.pageSize,
      defaultCurrent = _ref.defaultCurrent,
      defaultPageSize = _ref.defaultPageSize;
    var tCurrent = current || defaultCurrent;
    var tPageSize = pageSize || defaultPageSize;
    if (tPageSize && tCurrent) {
      return tPageSize * (tCurrent - 1) + rowIndex + 1;
    }
    return rowIndex + 1;
  }
  if (isFunction.isFunction(col.cell)) {
    return col.cell(params);
  }
  if (isFunction.isFunction(col.render)) {
    return col.render(_objectSpread(_objectSpread({}, params), {}, {
      type: "cell"
    }));
  }
  var r = col.cell || col.render || get.get(row, col.colKey);
  if (![void 0, "", null].includes(r)) return r;
  if (extra !== null && extra !== void 0 && extra.cellEmptyContent) return extra.cellEmptyContent;
  return r;
}
function renderEllipsisCell(cellParams, params) {
  var cellNode = params.cellNode,
    tableElm = params.tableElm,
    columnLength = params.columnLength,
    classPrefix = params.classPrefix,
    overlayClassName = params.overlayClassName;
  var col = cellParams.col,
    colIndex = cellParams.colIndex;
  var content = isFunction.isFunction(col.ellipsis) ? col.ellipsis(cellParams) : void 0;
  if (_typeof._typeof(col.ellipsis) === "object" && "content" in col.ellipsis && isFunction.isFunction(col.ellipsis.content)) {
    content = col.ellipsis.content(cellParams);
  }
  var tooltipProps = {};
  if (_typeof._typeof(col.ellipsis) === "object") {
    tooltipProps = "props" in col.ellipsis ? col.ellipsis.props : col.ellipsis || void 0;
  }
  var tableElement = tableElm;
  var placement = colIndex === 0 ? "top-left" : "top";
  placement = colIndex === columnLength - 1 ? "top-right" : placement;
  return /* @__PURE__ */React__default["default"].createElement(table_Ellipsis["default"], {
    placement: placement,
    attach: tableElement ? function () {
      return tableElement;
    } : void 0,
    popupContent: content,
    tooltipProps: tooltipProps,
    overlayClassName: overlayClassName,
    classPrefix: classPrefix
  }, cellNode);
}
var Cell = function Cell(props) {
  var cellParams = props.cellParams,
    tableClassNames = props.tableClassNames,
    tableRef = props.tableRef,
    columnLength = props.columnLength,
    classPrefix = props.classPrefix,
    overlayClassName = props.overlayClassName,
    pagination = props.pagination;
  var col = cellParams.col,
    colIndex = cellParams.colIndex,
    rowIndex = cellParams.rowIndex;
  var cellSpans = props.cellSpans,
    dataLength = props.dataLength,
    rowAndColFixedPosition = props.rowAndColFixedPosition,
    cellEmptyContent = props.cellEmptyContent,
    rowspanAndColspan = props.rowspanAndColspan,
    onClick = props.onClick;
  var tableColFixedClasses = tableClassNames.tableColFixedClasses,
    tdEllipsisClass = tableClassNames.tdEllipsisClass,
    tableBaseClass = tableClassNames.tableBaseClass,
    tdAlignClasses = tableClassNames.tdAlignClasses,
    tableDraggableClasses = tableClassNames.tableDraggableClasses;
  var cellNode = renderCell(cellParams, {
    cellEmptyContent: cellEmptyContent,
    pagination: pagination
  });
  var tdStyles = table_hooks_useFixed.getColumnFixedStyles(col, colIndex, rowAndColFixedPosition, tableColFixedClasses);
  var customClasses = table_utils.formatClassNames(col.className, _objectSpread(_objectSpread({}, cellParams), {}, {
    type: "td"
  }));
  var classes = [tdStyles.classes, customClasses, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, tdEllipsisClass, col.ellipsis), tableBaseClass.tdLastRow, rowIndex + cellSpans.rowspan === dataLength), tableBaseClass.tdFirstCol, colIndex === 0 && rowspanAndColspan), tdAlignClasses[col.align], col.align && col.align !== "left"), tableDraggableClasses.handle, col.colKey === "drag")];
  var normalAttrs = isFunction.isFunction(col.attrs) ? col.attrs(_objectSpread(_objectSpread({}, cellParams), {}, {
    type: "td"
  })) : col.attrs;
  var attrs = _objectSpread(_objectSpread({}, normalAttrs), {}, {
    rowSpan: cellSpans.rowspan,
    colSpan: cellSpans.colspan
  });
  return /* @__PURE__ */React__default["default"].createElement("td", _objectSpread(_objectSpread({
    key: col.colKey || colIndex,
    className: classNames__default["default"](classes),
    style: tdStyles.style
  }, attrs), {}, {
    onClick: onClick
  }), col.ellipsis ? renderEllipsisCell(cellParams, {
    cellNode: cellNode,
    tableElm: tableRef.current,
    columnLength: columnLength,
    classPrefix: classPrefix,
    overlayClassName: overlayClassName
  }) : cellNode);
};
Cell.displayName = "Cell";

exports["default"] = Cell;
exports.renderCell = renderCell;
