// tdesign-react/cjs/table/defaultProps.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var baseTableDefaultProps = {
  bordered: false,
  columns: [],
  data: [],
  disableDataPage: false,
  empty: "",
  footData: [],
  footerAffixedBottom: false,
  headerAffixedTop: false,
  hover: false,
  lazyLoad: false,
  loading: void 0,
  resizable: false,
  rowKey: "id",
  showHeader: true,
  size: "medium",
  stripe: false,
  tableLayout: "fixed",
  verticalAlign: "middle"
};
var primaryTableDefaultProps = {
  columnControllerVisible: void 0,
  columns: [],
  expandIcon: true,
  defaultExpandedRowKeys: [],
  multipleSort: false,
  reserveSelectedRowOnPaginate: true,
  defaultSelectedRowKeys: [],
  showSortColumnBgColor: false,
  sortOnRowDraggable: false
};
var enhancedTableDefaultProps = {
  defaultExpandedTreeNodes: [],
  rowKey: void 0
};

exports.baseTableDefaultProps = baseTableDefaultProps;
exports.enhancedTableDefaultProps = enhancedTableDefaultProps;
exports.primaryTableDefaultProps = primaryTableDefaultProps;
