// tdesign-react/esm/date-picker/defaultProps.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

var datePickerDefaultProps = {
  allowInput: false,
  borderless: false,
  clearable: false,
  defaultTime: "00:00:00",
  enableTimePicker: false,
  format: void 0,
  mode: "date",
  multiple: false,
  needConfirm: true,
  placeholder: void 0,
  presetsPlacement: "bottom",
  size: "medium",
  status: "default",
  defaultValue: ""
};
var dateRangePickerDefaultProps = {
  allowInput: false,
  borderless: false,
  cancelRangeSelectLimit: false,
  clearable: false,
  defaultTime: ["00:00:00", "23:59:59"],
  enableTimePicker: false,
  mode: "date",
  needConfirm: true,
  panelPreselection: true,
  presetsPlacement: "bottom",
  size: "medium",
  status: "default",
  defaultValue: []
};
var datePickerPanelDefaultProps = {
  defaultTime: "00:00:00"
};
var dateRangePickerPanelDefaultProps = {
  defaultTime: ["00:00:00", "23:59:59"]
};

export { datePickerDefaultProps, datePickerPanelDefaultProps, dateRangePickerDefaultProps, dateRangePickerPanelDefaultProps };
