// tdesign-react/cjs/sticky-tool/StickyTool.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-5481698f.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var stickyTool_StickyItem = require('./StickyItem.js');
var stickyTool_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-dab07be2.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-abe9dd70.js');
require('../_chunks/dep-75a87f7c.js');
require('dayjs');
require('../_chunks/dep-f0496dae.js');
require('../_chunks/dep-89f1315e.js');
require('../_chunks/dep-49b692b4.js');
require('../_chunks/dep-5dc79997.js');
require('../_chunks/dep-350e0ec1.js');
require('../_chunks/dep-cef99009.js');
require('../_chunks/dep-a241cf5a.js');
require('../_chunks/dep-7a32c861.js');
require('../_chunks/dep-d4fd8ae6.js');
require('../_chunks/dep-de258415.js');
require('../_chunks/dep-b20ed9a4.js');
require('../_chunks/dep-ea572de0.js');
require('../_chunks/dep-ac56ab8f.js');
require('../_chunks/dep-e3f45bee.js');
require('../_chunks/dep-7ffb0d81.js');
require('../_chunks/dep-d02e27a7.js');
require('../_chunks/dep-dd8855d3.js');
require('../_chunks/dep-c2bb7040.js');
require('../_chunks/dep-ffef71d0.js');
require('../_chunks/dep-53b02ac8.js');
require('../_chunks/dep-54d84720.js');
require('../_chunks/dep-da766de6.js');
require('hoist-non-react-statics');
require('../popup/index.js');
require('../popup/Popup.js');
require('../_chunks/dep-b20af383.js');
require('../_chunks/dep-7225af62.js');
require('react-transition-group');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-2a4a8f70.js');
require('../_chunks/dep-ba13d061.js');
require('../_chunks/dep-1e0bc473.js');
require('../_chunks/dep-fff57180.js');
require('../_chunks/dep-ac9a26ed.js');
require('../hooks/useAnimation.js');
require('../common/Portal.js');
require('react-dom');
require('../_util/dom.js');
require('raf');
require('../_util/easing.js');
require('../_chunks/dep-b2af1b16.js');
require('../hooks/useLayoutEffect.js');
require('../popup/hooks/useTrigger.js');
require('react-is');
require('../_util/ref.js');
require('../_util/isFragment.js');
require('../_util/composeRefs.js');
require('../popup/utils/transition.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-3e4c9d3d.js');
require('../_chunks/dep-9bee38bb.js');
require('../_chunks/dep-0ba932e9.js');
require('../_chunks/dep-b686b3fb.js');
require('../_chunks/dep-307d0db8.js');
require('../_chunks/dep-b3648f6a.js');
require('../_chunks/dep-62fd65d7.js');
require('../_chunks/dep-62978241.js');
require('../_chunks/dep-a8e37273.js');
require('../_chunks/dep-ae011087.js');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../hooks/useAttach.js');
require('../hooks/usePopper.js');
require('../_chunks/dep-b691f3a6.js');
require('@popperjs/core');
require('react-fast-compare');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-2b39f9ea.js');
require('../_util/react-render.js');
require('react-dom/client');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-040a2db4.js');
require('../_chunks/dep-641538cb.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StickyTool = _util_forwardRefWithStatics["default"](function (props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, stickyTool_defaultProps.stickyToolDefaultProps),
    style = _useDefaultProps.style,
    className = _useDefaultProps.className,
    children = _useDefaultProps.children,
    width = _useDefaultProps.width,
    type = _useDefaultProps.type,
    shape = _useDefaultProps.shape,
    placement = _useDefaultProps.placement,
    offset = _useDefaultProps.offset,
    popupProps = _useDefaultProps.popupProps,
    list = _useDefaultProps.list,
    onClick = _useDefaultProps.onClick,
    onHover = _useDefaultProps.onHover;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var styles = React.useMemo(function () {
    var position = offset ? [80, 24] : ["80px", "24px"];
    offset === null || offset === void 0 || offset.forEach(function (item, index) {
      position[index] = isNaN(Number(item)) ? "calc( ".concat(position[index], "px + ").concat(item, ")") : "".concat(position[index] + item, "px");
    });
    var styles2 = _objectSpread({}, style);
    placement.split("-").forEach(function (item, index) {
      if (item !== "center") {
        styles2[item] = position[index];
      } else {
        styles2.top = "50%";
        styles2.transform = "translate(0, -50%)";
      }
    });
    if (width) {
      styles2.width = typeof width === "number" ? "".concat(width, "px") : width;
    }
    return styles2;
  }, [offset, placement, width, style]);
  var handleClick = React.useCallback(function (context) {
    onClick === null || onClick === void 0 || onClick(context);
  }, [onClick]);
  var handleHover = React.useCallback(function (context) {
    onHover === null || onHover === void 0 || onHover(context);
  }, [onHover]);
  var stickyItemList = React.useMemo(function () {
    if (list !== null && list !== void 0 && list.length) {
      return list.map(function (item, index) {
        var itemProps = _objectSpread(_objectSpread({}, item), {}, {
          type: type,
          shape: shape,
          placement: placement,
          basePopupProps: popupProps,
          baseWidth: styles.width,
          onClick: handleClick,
          onHover: handleHover
        });
        return /* @__PURE__ */React__default["default"].createElement(stickyTool_StickyItem["default"], _objectSpread({
          key: index
        }, itemProps));
      });
    }
    var childrenList = React__default["default"].Children.toArray(children);
    return childrenList.map(function (child) {
      var itemProps = _objectSpread(_objectSpread({}, child.props), {}, {
        type: type,
        shape: shape,
        placement: placement,
        basePopupProps: popupProps,
        baseWidth: styles.width,
        onClick: handleClick,
        onHover: handleHover
      });
      return /*#__PURE__*/React__default["default"].cloneElement(child, _objectSpread({}, itemProps));
    });
  }, [list, children, type, shape, placement, popupProps, styles, handleClick, handleHover]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    style: styles,
    className: classNames__default["default"]("".concat(classPrefix, "-sticky-tool"), "".concat(classPrefix, "-sticky-tool--").concat(shape), className)
  }, stickyItemList);
}, {
  StickyItem: stickyTool_StickyItem["default"]
});
StickyTool.displayName = "StickyTool";

exports["default"] = StickyTool;
