// tdesign-react/lib/list/ListItemMeta.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

import React, { forwardRef } from 'react';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-daac9b4e.js';
import '../_chunks/dep-2b61bedd.js';
import 'lodash-es';
import '../_chunks/dep-d2abf305.js';
import '../_chunks/dep-d1906e4e.js';
import 'dayjs';

var ListItemMeta = /*#__PURE__*/forwardRef(function (props, ref) {
  var title = props.title,
    image = props.image,
    description = props.description,
    className = props.className,
    style = props.style;
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var renderAvatar = function renderAvatar() {
    if (image && typeof image === "string") {
      return /* @__PURE__ */React.createElement("div", {
        className: "".concat(classPrefix, "-list-item__meta-avatar")
      }, /* @__PURE__ */React.createElement("img", {
        src: image,
        alt: ""
      }));
    }
    return /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-list-item__meta-avatar")
    }, image);
  };
  return /* @__PURE__ */React.createElement("div", {
    ref: ref,
    className: classNames("".concat(classPrefix, "-list-item__meta"), className),
    style: style
  }, image && renderAvatar(), /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-list-item__meta-content")
  }, /* @__PURE__ */React.createElement("h3", {
    className: "".concat(classPrefix, "-list-item__meta-title")
  }, title), /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-list-item__meta-description")
  }, typeof description === "string" ? /* @__PURE__ */React.createElement("p", null, description) : description)));
});
ListItemMeta.displayName = "ListItemMeta";

export { ListItemMeta as default };
