// tdesign-react/es/table/hooks/usePagination.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-daac9b4e.js';
import { _ as _toConsumableArray } from '../../_chunks/dep-385f8c20.js';
import { _ as _slicedToArray } from '../../_chunks/dep-17424654.js';
import React, { useState, useCallback, useEffect } from 'react';
import useConfig from '../../hooks/useConfig.js';
import { Pagination } from '../../pagination/index.js';
import '../../_chunks/dep-2b61bedd.js';
import '../../_chunks/dep-73b0b578.js';
import '../../config-provider/ConfigContext.js';
import 'lodash-es';
import '../../locale/zh_CN.js';
import '../../_chunks/dep-34028257.js';
import 'dayjs';
import '../../_chunks/dep-55f7c5cf.js';
import '../../pagination/Pagination.js';
import '../../_chunks/dep-b209b5b7.js';
import 'classnames';
import '../../_util/noop.js';
import '../../hooks/useControlled.js';
import '../../select/index.js';
import '../../select/base/Select.js';
import '../../_util/helper.js';
import '../../locale/LocalReceiver.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../config-provider/type.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../select/util/helper.js';
import '../../select/base/OptionGroup.js';
import '../../select/defaultProps.js';
import '../../hooks/useDefaultProps.js';
import '../../select/base/Option.js';
import '../../hooks/useDomRefCallback.js';
import '../../hooks/useRipple.js';
import '../../hooks/useAnimation.js';
import '../../_util/setStyle.js';
import '../../_util/dom.js';
import 'raf';
import '../../_util/easing.js';
import '../../common/FakeArrow.js';
import '../../loading/index.js';
import '../../loading/Loading.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../hooks/useLayoutEffect.js';
import '../../_chunks/dep-6be0c78e.js';
import '../../_chunks/dep-a0aea470.js';
import '../../loading/defaultProps.js';
import '../../loading/plugin.js';
import '../../_util/react-render.js';
import '../../_chunks/dep-4e82f065.js';
import 'react-dom/client';
import '../../common/PluginContainer.js';
import '../../loading/style/css.js';
import '../../loading/type.js';
import '../../select-input/index.js';
import '../../select-input/SelectInput.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import '../../popup/hooks/useTrigger.js';
import 'react-is';
import '../../_util/ref.js';
import '../../_util/isFragment.js';
import '../../_util/composeRefs.js';
import '../../popup/utils/transition.js';
import '../../hooks/useMutationObserver.js';
import '../../hooks/useLatest.js';
import '../../hooks/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../hooks/useAttach.js';
import '../../hooks/usePopper.js';
import '@popperjs/core';
import 'react-fast-compare';
import '../../popup/PopupPlugin.js';
import '../../popup/style/css.js';
import '../../popup/type.js';
import '../../select-input/useSingle.js';
import '../../input/index.js';
import '../../input/Input.js';
import 'tdesign-icons-react';
import '../../hooks/useGlobalIcon.js';
import '../../input/InputGroup.js';
import '../../input/defaultProps.js';
import '../../_util/parseTNode.js';
import '../../_chunks/dep-4d98bbe1.js';
import '../../_chunks/dep-abe679d5.js';
import '../../input/useLengthLimit.js';
import '../../input/style/css.js';
import '../../input/type.js';
import '../../select-input/useMultiple.js';
import '../../tag-input/index.js';
import '../../tag-input/TagInput.js';
import '../../hooks/useDragSorter.js';
import '../../hooks/useEventCallback.js';
import '../../tag-input/useTagScroll.js';
import '../../tag-input/useTagList.js';
import '../../tag/index.js';
import '../../tag/Tag.js';
import 'tinycolor2';
import '../../tag/defaultProps.js';
import '../../tag/CheckTag.js';
import '../../_chunks/dep-3c6e2298.js';
import '../../tag/CheckTagGroup.js';
import '../../tag/style/css.js';
import '../../tag/type.js';
import '../../tag-input/useHover.js';
import '../../tag-input/defaultProps.js';
import '../../tag-input/style/css.js';
import '../../tag-input/type.js';
import '../../select-input/useOverlayInnerStyle.js';
import '../../select-input/defaultProps.js';
import '../../select-input/style/css.js';
import '../../select-input/type.js';
import '../../select/base/PopupContent.js';
import '../../select/hooks/usePanelVirtualScroll.js';
import '../../hooks/useVirtualScroll.js';
import '../../hooks/useResizeObserver.js';
import '../../select/hooks/useOptions.js';
import '../../select/style/css.js';
import '../../select/type.js';
import '../../input-number/index.js';
import '../../input-number/InputNumber.js';
import '../../button/index.js';
import '../../button/Button.js';
import '../../button/defaultProps.js';
import '../../button/style/css.js';
import '../../button/type.js';
import '../../input-number/useInputNumber.js';
import '../../_chunks/dep-3e196638.js';
import '../../hooks/useCommonClassName.js';
import '../../input-number/defaultProps.js';
import '../../input-number/style/css.js';
import '../../input-number/type.js';
import '../../input-adornment/index.js';
import '../../input-adornment/InputAdornment.js';
import '../../input-adornment/style/css.js';
import '../../input-adornment/type.js';
import '../../pagination/hooks/useBoundaryJumper.js';
import '../../pagination/hooks/usePrevNextJumper.js';
import '../../pagination/hooks/usePageNumber.js';
import '../../pagination/hooks/useTotal.js';
import '../../pagination/validators.js';
import '../../pagination/defaultProps.js';
import '../../pagination/PaginationMini.js';
import '../../pagination/style/css.js';
import '../../pagination/type.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function usePagination(props) {
  var pagination = props.pagination,
    data = props.data,
    disableDataPage = props.disableDataPage;
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var _useState = useState(props.pagination),
    _useState2 = _slicedToArray(_useState, 2),
    innerPagination = _useState2[0],
    setInnerPagination = _useState2[1];
  var _useState3 = useState([]),
    _useState4 = _slicedToArray(_useState3, 2),
    dataSource = _useState4[0],
    setDataSource = _useState4[1];
  var _useState5 = useState(false),
    _useState6 = _slicedToArray(_useState5, 2),
    isPaginateData = _useState6[0],
    setIsPaginateData = _useState6[1];
  var updateDataSourceAndPaginate = useCallback(function () {
    var current = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    var pageSize = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 10;
    var isPaginateData2 = Boolean(!disableDataPage && data.length > pageSize);
    setIsPaginateData(isPaginateData2);
    var newData = [];
    if (isPaginateData2) {
      var start = (current - 1) * pageSize;
      var end = current * pageSize;
      newData = _toConsumableArray(data.slice(start, end));
    } else {
      newData = data;
    }
    setDataSource(newData);
    return newData;
  }, [data, disableDataPage]);
  useEffect(function () {
    if (!pagination) {
      setIsPaginateData(false);
    }
  }, [pagination]);
  useEffect(function () {
    var _pagination$pageSize;
    if (!pagination || !pagination.current) return;
    var current = pagination === null || pagination === void 0 ? void 0 : pagination.current,
      pageSize = (_pagination$pageSize = pagination === null || pagination === void 0 ? void 0 : pagination.pageSize) !== null && _pagination$pageSize !== void 0 ? _pagination$pageSize : 10;
    updateDataSourceAndPaginate(current, pageSize);
    setInnerPagination({
      current: current,
      pageSize: pageSize
    });
  }, [pagination, updateDataSourceAndPaginate]);
  useEffect(function () {
    var _innerPagination$curr, _innerPagination$page;
    if (!pagination || !pagination.defaultCurrent) return;
    var isControlled = Boolean(pagination.current);
    if (isControlled) return;
    updateDataSourceAndPaginate((_innerPagination$curr = innerPagination.current) !== null && _innerPagination$curr !== void 0 ? _innerPagination$curr : pagination.defaultCurrent, (_innerPagination$page = innerPagination.pageSize) !== null && _innerPagination$page !== void 0 ? _innerPagination$page : pagination.defaultPageSize);
  }, [updateDataSourceAndPaginate]);
  var renderPagination = function renderPagination() {
    if (!pagination) return null;
    return /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-table__pagination")
    }, /* @__PURE__ */React.createElement(Pagination, _objectSpread(_objectSpread({}, pagination), {}, {
      onChange: function onChange(pageInfo) {
        var _props$pagination, _props$pagination$onC, _props$onPageChange;
        (_props$pagination = props.pagination) === null || _props$pagination === void 0 || (_props$pagination$onC = _props$pagination.onChange) === null || _props$pagination$onC === void 0 || _props$pagination$onC.call(_props$pagination, pageInfo);
        setInnerPagination(pageInfo);
        var newData = updateDataSourceAndPaginate(pageInfo.current, pageInfo.pageSize);
        (_props$onPageChange = props.onPageChange) === null || _props$onPageChange === void 0 || _props$onPageChange.call(props, pageInfo, newData);
      }
    })));
  };
  return {
    isPaginateData: isPaginateData,
    dataSource: dataSource,
    innerPagination: innerPagination,
    renderPagination: renderPagination
  };
}

export { usePagination as default };
