// tdesign-react/cjs/empty/Empty.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-b20af383.js');
var defineProperty = require('../_chunks/dep-5481698f.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var empty_defaultProps = require('./defaultProps.js');
var image_index = require('../image/index.js');
var empty_assets_MaintenanceSvg = require('./assets/MaintenanceSvg.js');
var empty_assets_NetworkErrorSvg = require('./assets/NetworkErrorSvg.js');
var empty_assets_EmptySvg = require('./assets/EmptySvg.js');
var empty_assets_FailSvg = require('./assets/FailSvg.js');
var empty_assets_SuccessSvg = require('./assets/SuccessSvg.js');
var isObject = require('../_chunks/dep-de258415.js');
var isString = require('../_chunks/dep-b2af1b16.js');
require('../_chunks/dep-7225af62.js');
require('../_chunks/dep-dab07be2.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-abe9dd70.js');
require('../_chunks/dep-75a87f7c.js');
require('dayjs');
require('../_chunks/dep-f0496dae.js');
require('../_chunks/dep-89f1315e.js');
require('../_chunks/dep-49b692b4.js');
require('../_chunks/dep-5dc79997.js');
require('../_chunks/dep-350e0ec1.js');
require('../_chunks/dep-cef99009.js');
require('../_chunks/dep-a241cf5a.js');
require('../_chunks/dep-7a32c861.js');
require('../_chunks/dep-d4fd8ae6.js');
require('../_chunks/dep-b20ed9a4.js');
require('../_chunks/dep-ea572de0.js');
require('../_chunks/dep-ac56ab8f.js');
require('../_chunks/dep-e3f45bee.js');
require('../_chunks/dep-7ffb0d81.js');
require('../_chunks/dep-d02e27a7.js');
require('../_chunks/dep-dd8855d3.js');
require('../_chunks/dep-c2bb7040.js');
require('../_chunks/dep-ffef71d0.js');
require('../_chunks/dep-53b02ac8.js');
require('../_chunks/dep-54d84720.js');
require('../_chunks/dep-da766de6.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-040a2db4.js');
require('../_chunks/dep-641538cb.js');
require('../_chunks/dep-b686b3fb.js');
require('../_chunks/dep-307d0db8.js');
require('../_chunks/dep-b3648f6a.js');
require('../_chunks/dep-62fd65d7.js');
require('../_chunks/dep-62978241.js');
require('../image/Image.js');
require('../_chunks/dep-d19a672c.js');
require('tdesign-icons-react');
require('../_chunks/dep-e3615e7d.js');
require('../image/defaultProps.js');
require('../space/index.js');
require('../space/Space.js');
require('react-is');
require('../_chunks/dep-be7119b7.js');
require('../_chunks/dep-b691f3a6.js');
require('../_chunks/dep-ca323c2b.js');
require('../space/defaultProps.js');
require('../hooks/useGlobalIcon.js');
require('../hooks/useImagePreviewUrl.js');
require('../_chunks/dep-e831f387.js');
require('../_chunks/dep-2c9a43ee.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function getImageIns(data) {
  var result = data;
  if (/*#__PURE__*/React.isValidElement(data)) {
    result = data;
  } else if (isObject.isObject(data)) {
    result = /* @__PURE__ */React__default["default"].createElement(image_index.Image, _objectSpread({}, data));
  } else if (isString.isString(data)) {
    result = /* @__PURE__ */React__default["default"].createElement(image_index.Image, {
      src: data
    });
  }
  return data ? result : null;
}
var Empty = function Empty(props) {
  var _defaultMaps$type, _typeImageProps$image, _typeImageProps$title;
  var _useDefaultProps = hooks_useDefaultProps["default"](props, empty_defaultProps.emptyDefaultProps),
    propsImage = _useDefaultProps.image,
    imageStyle = _useDefaultProps.imageStyle,
    propsDescription = _useDefaultProps.description,
    propsTitle = _useDefaultProps.title,
    type = _useDefaultProps.type,
    action = _useDefaultProps.action,
    style = _useDefaultProps.style,
    className = _useDefaultProps.className,
    _useDefaultProps$size = _useDefaultProps.size,
    size = _useDefaultProps$size === void 0 ? "medium" : _useDefaultProps$size;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix,
    empty = _useConfig.empty;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("empty"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var defaultMaps = {
    maintenance: {
      image: empty.image.maintenance || /* @__PURE__ */React__default["default"].createElement(empty_assets_MaintenanceSvg["default"], null),
      title: empty.titleText.maintenance || t(local.titleText.maintenance)
    },
    success: {
      image: empty.image.success || /* @__PURE__ */React__default["default"].createElement(empty_assets_SuccessSvg["default"], null),
      title: empty.titleText.success || t(local.titleText.success)
    },
    fail: {
      image: empty.image.fail || /* @__PURE__ */React__default["default"].createElement(empty_assets_FailSvg["default"], null),
      title: empty.titleText.fail || t(local.titleText.fail)
    },
    "network-error": {
      image: empty.image.networkError || /* @__PURE__ */React__default["default"].createElement(empty_assets_NetworkErrorSvg["default"], null),
      title: empty.titleText.networkError || t(local.titleText.networkError)
    },
    empty: {
      image: empty.image.empty || /* @__PURE__ */React__default["default"].createElement(empty_assets_EmptySvg["default"], null),
      title: empty.titleText.empty || t(local.titleText.empty)
    }
  };
  var defaultSize = {
    small: "".concat(classPrefix, "-size-s"),
    medium: "".concat(classPrefix, "-size"),
    large: "".concat(classPrefix, "-size-l")
  };
  var prefix = "".concat(classPrefix, "-empty");
  var emptyClasses = classNames__default["default"](prefix, className, defaultSize[size]);
  var titleClasses = classNames__default["default"]("".concat(prefix, "__title"));
  var imageClasses = classNames__default["default"]("".concat(prefix, "__image"));
  var descriptionClasses = classNames__default["default"]("".concat(prefix, "__description"));
  var actionCls = classNames__default["default"]("".concat(prefix, "__action"));
  var typeImageProps = (_defaultMaps$type = defaultMaps[type]) !== null && _defaultMaps$type !== void 0 ? _defaultMaps$type : null;
  var _image$title$descript = {
      image: propsImage ? propsImage : (_typeImageProps$image = typeImageProps === null || typeImageProps === void 0 ? void 0 : typeImageProps.image) !== null && _typeImageProps$image !== void 0 ? _typeImageProps$image : null,
      title: propsTitle ? propsTitle : (_typeImageProps$title = typeImageProps === null || typeImageProps === void 0 ? void 0 : typeImageProps.title) !== null && _typeImageProps$title !== void 0 ? _typeImageProps$title : null,
      description: propsDescription
    },
    image = _image$title$descript.image,
    description = _image$title$descript.description,
    title = _image$title$descript.title;
  function renderTitle() {
    if (!title) {
      return null;
    }
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: titleClasses
    }, title);
  }
  function renderDescription() {
    if (!description) {
      return null;
    }
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: descriptionClasses
    }, description);
  }
  var imageContent = getImageIns(image);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: emptyClasses,
    style: style
  }, imageContent ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: imageClasses,
    style: imageStyle
  }, imageContent) : null, renderTitle(), renderDescription(), action ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: actionCls
  }, action) : null);
};

exports["default"] = Empty;
