// tdesign-react/es/menu/hooks/useMenuContext.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

import { _ as _toConsumableArray } from '../../_chunks/dep-385f8c20.js';
import { _ as _defineProperty } from '../../_chunks/dep-daac9b4e.js';
import { _ as _slicedToArray } from '../../_chunks/dep-17424654.js';
import { noop } from 'lodash-es';
import { useState } from 'react';
import checkSubMenuChildExpanded from '../_util/checkSubMenuChildExpanded.js';
import '../../_chunks/dep-73b0b578.js';
import '../../_chunks/dep-2b61bedd.js';
import '../_util/type.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function useMenuContext(_ref) {
  var defaultValue = _ref.defaultValue,
    defaultExpanded = _ref.defaultExpanded,
    expandMutex = _ref.expandMutex,
    children = _ref.children,
    _ref$onExpand = _ref.onExpand,
    onExpand = _ref$onExpand === void 0 ? noop : _ref$onExpand,
    _ref$onChange = _ref.onChange,
    onChange = _ref$onChange === void 0 ? noop : _ref$onChange,
    value = _ref.value,
    expanded = _ref.expanded,
    collapsed = _ref.collapsed,
    expandType = _ref.expandType,
    theme = _ref.theme,
    mode = _ref.mode;
  var _useState = useState({
      active: defaultValue,
      expanded: defaultExpanded
    }),
    _useState2 = _slicedToArray(_useState, 2),
    state = _useState2[0],
    setState = _useState2[1];
  var setStateValue = function setStateValue(menuState) {
    return setState(_objectSpread(_objectSpread({}, state), menuState));
  };
  var handleExpandChange = function handleExpandChange(value2, expanded2) {
    var nextExpand = [];
    var index = expanded2.indexOf(value2);
    if (index > -1) {
      if (expandMutex) {
        nextExpand = expanded2.slice(0, index);
      } else {
        nextExpand = expanded2.filter(function (item) {
          return item !== value2;
        });
      }
    } else if (expandMutex) {
      nextExpand = checkSubMenuChildExpanded(children, expanded2, value2);
    } else {
      nextExpand = [].concat(_toConsumableArray(expanded2), [value2]);
    }
    onExpand(nextExpand);
    setState({
      expanded: nextExpand
    });
  };
  return {
    value: {
      onExpand: handleExpandChange,
      onChange: onChange,
      active: value || state.active,
      expanded: expanded || state.expanded,
      collapsed: collapsed || state.collapsed,
      setState: setStateValue,
      expandMutex: expandMutex,
      expandType: collapsed ? "popup" : expandType,
      mode: mode,
      theme: theme
    }
  };
}

export { useMenuContext as default };
