// tdesign-react/cjs/typography/ellipsis/Ellipsis.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-5481698f.js');
var slicedToArray = require('../../_chunks/dep-b20af383.js');
var objectWithoutProperties = require('../../_chunks/dep-d19a672c.js');
var React = require('react');
var classNames = require('classnames');
var typography_ellipsis_Truncate = require('./Truncate.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
require('../../_chunks/dep-dab07be2.js');
require('../../_chunks/dep-7225af62.js');
require('../../_chunks/dep-e711efe7.js');
require('react-is');
require('../../_chunks/dep-be9ef47b.js');
require('../../_chunks/dep-fff57180.js');
require('../../_chunks/dep-5dc79997.js');
require('../../_chunks/dep-cef99009.js');
require('../../_chunks/dep-ac9a26ed.js');
require('../../_chunks/dep-350e0ec1.js');
require('../../_chunks/dep-641538cb.js');
require('../../_chunks/dep-89f1315e.js');
require('../../_chunks/dep-49b692b4.js');
require('../../_chunks/dep-a241cf5a.js');
require('../../_chunks/dep-7a32c861.js');
require('../../_chunks/dep-d4fd8ae6.js');
require('../../_chunks/dep-de258415.js');
require('../../_chunks/dep-b20ed9a4.js');
require('../../_chunks/dep-ea572de0.js');
require('../../_chunks/dep-ac56ab8f.js');
require('../../_chunks/dep-e3f45bee.js');
require('../../_chunks/dep-7ffb0d81.js');
require('../../_chunks/dep-d02e27a7.js');
require('../../_chunks/dep-dd8855d3.js');
require('../../_chunks/dep-c2bb7040.js');
require('../../_chunks/dep-b686b3fb.js');
require('../../_chunks/dep-307d0db8.js');
require('../../_chunks/dep-b3648f6a.js');
require('../../_chunks/dep-62fd65d7.js');
require('../../_chunks/dep-62978241.js');
require('../../_chunks/dep-29a8a75a.js');
require('../../_chunks/dep-8b483a12.js');
require('../../_chunks/dep-1e0bc473.js');
require('../../_chunks/dep-54d84720.js');
require('../../_chunks/dep-c74cf9fd.js');
require('../../_chunks/dep-781be47c.js');
require('../../_chunks/dep-53b02ac8.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-abe9dd70.js');
require('../../_chunks/dep-75a87f7c.js');
require('dayjs');
require('../../_chunks/dep-f0496dae.js');
require('../../_chunks/dep-ffef71d0.js');
require('../../_chunks/dep-da766de6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["className", "children", "lines", "ellipsisClassName", "ellipsisPrefix", "onToggleExpand", "width", "onTruncate", "component", "collapsible", "expandable", "more", "less"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Ellipsis = function Ellipsis(_ref) {
  var className = _ref.className,
    children = _ref.children,
    _ref$lines = _ref.lines,
    lines = _ref$lines === void 0 ? 1 : _ref$lines,
    ellipsisClassName = _ref.ellipsisClassName,
    _ref$ellipsisPrefix = _ref.ellipsisPrefix,
    ellipsisPrefix = _ref$ellipsisPrefix === void 0 ? "..." : _ref$ellipsisPrefix,
    onToggleExpand = _ref.onToggleExpand,
    _ref$width = _ref.width,
    width = _ref$width === void 0 ? 0 : _ref$width,
    onTruncate = _ref.onTruncate,
    _ref$component = _ref.component,
    Component = _ref$component === void 0 ? "div" : _ref$component,
    _ref$collapsible = _ref.collapsible,
    collapsible = _ref$collapsible === void 0 ? false : _ref$collapsible,
    _ref$expandable = _ref.expandable,
    expandable = _ref$expandable === void 0 ? false : _ref$expandable,
    more = _ref.more,
    less = _ref.less,
    rest = objectWithoutProperties._objectWithoutProperties(_ref, _excluded);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var symbolClassName = ellipsisClassName || "".concat(classPrefix, "-typography-ellipsis-symbol");
  var isMountRef = React.useRef(false);
  React.useEffect(function () {
    isMountRef.current = true;
  }, []);
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isExpanded = _useState2[0],
    setIsExpanded = _useState2[1];
  var handleToggleExpand = function handleToggleExpand(e) {
    if (!expandable) return;
    if (isMountRef.current) {
      setIsExpanded(!isExpanded);
      onToggleExpand === null || onToggleExpand === void 0 || onToggleExpand(!isExpanded, e);
    }
  };
  var truncateRef = React.useRef(null);
  var _useState3 = React.useState(false),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    isTruncated = _useState4[0],
    setTruncated = _useState4[1];
  var handleTruncate = function handleTruncate(truncated) {
    if (isMountRef.current && truncated !== isTruncated) {
      setTruncated(truncated);
      if (truncated && truncateRef.current) {
        var _truncateRef$current$, _truncateRef$current;
        (_truncateRef$current$ = (_truncateRef$current = truncateRef.current).onResize) === null || _truncateRef$current$ === void 0 || _truncateRef$current$.call(_truncateRef$current);
      }
      onTruncate === null || onTruncate === void 0 || onTruncate(truncated);
    }
  };
  var componentProps = _objectSpread({
    className: className
  }, rest);
  return /* @__PURE__ */React__default["default"].createElement(Component, _objectSpread({}, componentProps), /* @__PURE__ */React__default["default"].createElement(typography_ellipsis_Truncate["default"], {
    width: width,
    lines: !isExpanded && lines,
    className: "".concat(classPrefix, "-typography-ellipsis"),
    ellipsis: /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-typography-ellipsis-symbol-wrapper")
    }, ellipsisPrefix, /* @__PURE__ */React__default["default"].createElement("span", {
      className: symbolClassName,
      onClick: handleToggleExpand
    }, more)),
    onTruncate: handleTruncate,
    ref: truncateRef,
    lineClassName: "".concat(classPrefix, "-typography-ellipsis-line")
  }, children), !isTruncated && collapsible && isExpanded && /* @__PURE__ */React__default["default"].createElement("span", {
    className: classNames__default["default"](symbolClassName, "".concat(symbolClassName, "--expanded")),
    onClick: handleToggleExpand
  }, less));
};

exports["default"] = Ellipsis;
