// tdesign-react/cjs/typography/ellipsis/useEllipsis.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-5481698f.js');
var slicedToArray = require('../../_chunks/dep-b20af383.js');
var React = require('react');
var tooltip_index = require('../../tooltip/index.js');
var locale_LocalReceiver = require('../../locale/LocalReceiver.js');
var isFunction = require('../../_chunks/dep-d4fd8ae6.js');
require('../../_chunks/dep-dab07be2.js');
require('../../_chunks/dep-7225af62.js');
require('../../tooltip/Tooltip.js');
require('../../_chunks/dep-d19a672c.js');
require('classnames');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('../../hooks/useControlled.js');
require('../../_util/noop.js');
require('../../_chunks/dep-2a4a8f70.js');
require('../../_chunks/dep-ba13d061.js');
require('../../_chunks/dep-1e0bc473.js');
require('../../_chunks/dep-fff57180.js');
require('../../_chunks/dep-5dc79997.js');
require('../../_chunks/dep-cef99009.js');
require('../../_chunks/dep-ac9a26ed.js');
require('../../_chunks/dep-350e0ec1.js');
require('../../hooks/useAnimation.js');
require('../../hooks/useConfig.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-abe9dd70.js');
require('../../_chunks/dep-75a87f7c.js');
require('dayjs');
require('../../_chunks/dep-f0496dae.js');
require('../../_chunks/dep-89f1315e.js');
require('../../_chunks/dep-49b692b4.js');
require('../../_chunks/dep-a241cf5a.js');
require('../../_chunks/dep-7a32c861.js');
require('../../_chunks/dep-b20ed9a4.js');
require('../../_chunks/dep-ea572de0.js');
require('../../_chunks/dep-ac56ab8f.js');
require('../../_chunks/dep-e3f45bee.js');
require('../../_chunks/dep-de258415.js');
require('../../_chunks/dep-7ffb0d81.js');
require('../../_chunks/dep-d02e27a7.js');
require('../../_chunks/dep-dd8855d3.js');
require('../../_chunks/dep-c2bb7040.js');
require('../../_chunks/dep-ffef71d0.js');
require('../../_chunks/dep-53b02ac8.js');
require('../../_chunks/dep-54d84720.js');
require('../../_chunks/dep-da766de6.js');
require('../../common/Portal.js');
require('react-dom');
require('../../_util/dom.js');
require('raf');
require('../../_util/easing.js');
require('../../_chunks/dep-b2af1b16.js');
require('../../hooks/useLayoutEffect.js');
require('../../popup/hooks/useTrigger.js');
require('react-is');
require('../../_util/ref.js');
require('../../_util/isFragment.js');
require('../../_util/composeRefs.js');
require('../../popup/utils/transition.js');
require('../../hooks/useMutationObserver.js');
require('../../hooks/useLatest.js');
require('../../_chunks/dep-3e4c9d3d.js');
require('../../_chunks/dep-9bee38bb.js');
require('../../_chunks/dep-0ba932e9.js');
require('../../_chunks/dep-b686b3fb.js');
require('../../_chunks/dep-307d0db8.js');
require('../../_chunks/dep-b3648f6a.js');
require('../../_chunks/dep-62fd65d7.js');
require('../../_chunks/dep-62978241.js');
require('../../_chunks/dep-a8e37273.js');
require('../../_chunks/dep-ae011087.js');
require('../../hooks/useWindowSize.js');
require('../../popup/defaultProps.js');
require('../../hooks/useDefaultProps.js');
require('../../hooks/useAttach.js');
require('../../hooks/usePopper.js');
require('../../_chunks/dep-b691f3a6.js');
require('@popperjs/core');
require('react-fast-compare');
require('../../popup/PopupPlugin.js');
require('../../_chunks/dep-2b39f9ea.js');
require('../../_util/react-render.js');
require('react-dom/client');
require('../../common/PluginContainer.js');
require('../../config-provider/ConfigProvider.js');
require('../../_chunks/dep-040a2db4.js');
require('../../_chunks/dep-641538cb.js');
require('../../tooltip/defaultProps.js');
require('../../tooltip/TooltipLite.js');
require('../../hooks/useSwitch.js');
require('../../hooks/usePersistFn.js');
require('../../_chunks/dep-02e4e6c8.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function useEllipsis(ellipsis) {
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("typography"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var expandText = t(local.expandText);
  var collapseText = t(local.collapseText);
  var formattedEllipsis = {};
  if (ellipsis) {
    var _ellipsis$expandable, _ellipsis$collapsible;
    formattedEllipsis = ellipsis === true ? {
      row: 1,
      expandable: false,
      tooltipProps: null,
      suffix: function suffix() {
        return "";
      },
      collapsible: true
    } : {
      row: ellipsis.row || 1,
      expandable: (_ellipsis$expandable = ellipsis.expandable) !== null && _ellipsis$expandable !== void 0 ? _ellipsis$expandable : false,
      tooltipProps: ellipsis.tooltipProps || null,
      suffix: function suffix(_ref) {
        var expanded = _ref.expanded;
        return typeof (ellipsis === null || ellipsis === void 0 ? void 0 : ellipsis.suffix) === "function" ? ellipsis === null || ellipsis === void 0 ? void 0 : ellipsis.suffix({
          expanded: expanded
        }) : expanded ? collapseText : ellipsis !== null && ellipsis !== void 0 && ellipsis.expandable ? "".concat(expandText) : "...";
      },
      collapsible: (_ellipsis$collapsible = ellipsis === null || ellipsis === void 0 ? void 0 : ellipsis.collapsible) !== null && _ellipsis$collapsible !== void 0 ? _ellipsis$collapsible : false
    };
  }
  var _useState = React.useState(true),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isClamped = _useState2[0],
    setIsClamped = _useState2[1];
  var handleExpand = function handleExpand(expanded) {
    var _ellipsis$onExpand;
    if (typeof expanded !== "boolean") return;
    setIsClamped(!expanded);
    (_ellipsis$onExpand = ellipsis.onExpand) === null || _ellipsis$onExpand === void 0 || _ellipsis$onExpand.call(ellipsis, !expanded);
  };
  var getEllipsisSuffix = function getEllipsisSuffix() {
    var _formattedEllipsis$su, _formattedEllipsis, _formattedEllipsis2;
    var moreOrLess;
    if (isFunction.isFunction(formattedEllipsis.suffix)) moreOrLess = (_formattedEllipsis$su = (_formattedEllipsis = formattedEllipsis).suffix) === null || _formattedEllipsis$su === void 0 ? void 0 : _formattedEllipsis$su.call(_formattedEllipsis, {
      expanded: !isClamped
    });else moreOrLess = formattedEllipsis.suffix;
    if ((_formattedEllipsis2 = formattedEllipsis) !== null && _formattedEllipsis2 !== void 0 && _formattedEllipsis2.tooltipProps && !!moreOrLess) {
      var _formattedEllipsis3;
      return /* @__PURE__ */React__default["default"].createElement(tooltip_index.Tooltip, _objectSpread({}, (_formattedEllipsis3 = formattedEllipsis) === null || _formattedEllipsis3 === void 0 ? void 0 : _formattedEllipsis3.tooltipProps), moreOrLess);
    }
    return moreOrLess;
  };
  var getEllipsisPrefix = function getEllipsisPrefix() {
    var _formattedEllipsis$su2, _formattedEllipsis4, _formattedEllipsis5;
    var moreOrLess;
    if (isFunction.isFunction(formattedEllipsis.suffix)) moreOrLess = (_formattedEllipsis$su2 = (_formattedEllipsis4 = formattedEllipsis).suffix) === null || _formattedEllipsis$su2 === void 0 ? void 0 : _formattedEllipsis$su2.call(_formattedEllipsis4, {
      expanded: !isClamped
    });else moreOrLess = formattedEllipsis.suffix;
    if ((_formattedEllipsis5 = formattedEllipsis) !== null && _formattedEllipsis5 !== void 0 && _formattedEllipsis5.tooltipProps && !moreOrLess) {
      var _formattedEllipsis6;
      return /* @__PURE__ */React__default["default"].createElement(tooltip_index.Tooltip, _objectSpread({}, (_formattedEllipsis6 = formattedEllipsis) === null || _formattedEllipsis6 === void 0 ? void 0 : _formattedEllipsis6.tooltipProps), "...");
    }
    return "...";
  };
  var ellipsisProps = {
    lines: formattedEllipsis.row,
    ellipsisPrefix: getEllipsisPrefix(),
    more: getEllipsisSuffix(),
    less: getEllipsisSuffix(),
    onToggleExpand: handleExpand,
    expandable: formattedEllipsis.expandable,
    collapsible: formattedEllipsis.collapsible
  };
  return {
    ellipsisProps: ellipsisProps
  };
}

exports["default"] = useEllipsis;
