// tdesign-react/esm/select/hooks/useOptions.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-daac9b4e.js';
import { _ as _slicedToArray } from '../../_chunks/dep-17424654.js';
import React, { useState, useEffect } from 'react';
import { get } from 'lodash-es';
import { getValueToOption } from '../util/helper.js';
import Option from '../base/Option.js';
import OptionGroup from '../base/OptionGroup.js';
import '../../_chunks/dep-2b61bedd.js';
import '../../_chunks/dep-73b0b578.js';
import '../../_chunks/dep-385f8c20.js';
import 'classnames';
import '../../hooks/useConfig.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../_chunks/dep-34028257.js';
import 'dayjs';
import '../../_chunks/dep-55f7c5cf.js';
import '../defaultProps.js';
import '../../hooks/useDefaultProps.js';
import '../../hooks/useDomRefCallback.js';
import '../../hooks/useRipple.js';
import '../../hooks/useAnimation.js';
import '../../_util/setStyle.js';
import '../../_util/dom.js';
import 'raf';
import '../../_util/easing.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function UseOptions(keys, options, children, valueType, value, reserveKeyword) {
  var _useState = useState({}),
    _useState2 = _slicedToArray(_useState, 2),
    valueToOption = _useState2[0],
    setValueToOption = _useState2[1];
  var _useState3 = useState([]),
    _useState4 = _slicedToArray(_useState3, 2),
    currentOptions = _useState4[0],
    setCurrentOptions = _useState4[1];
  var _useState5 = useState([]),
    _useState6 = _slicedToArray(_useState5, 2),
    tmpPropOptions = _useState6[0],
    setTmpPropOptions = _useState6[1];
  var _useState7 = useState([]),
    _useState8 = _slicedToArray(_useState7, 2),
    selectedOptions = _useState8[0],
    setSelectedOptions = _useState8[1];
  useEffect(function () {
    var transformedOptions = options;
    var arrayChildren = React.Children.toArray(children);
    var optionChildren = arrayChildren.filter(function (v) {
      return v.type === Option || v.type === OptionGroup;
    });
    var isChildrenFilterable = arrayChildren.length > 0 && optionChildren.length === arrayChildren.length;
    if (reserveKeyword && currentOptions.length && isChildrenFilterable) return;
    if (isChildrenFilterable) {
      var _handlerOptionElement = function handlerOptionElement(v) {
        if (/*#__PURE__*/React.isValidElement(v)) {
          if (v.type === OptionGroup) {
            var _v$props$children;
            return _objectSpread(_objectSpread({}, v.props), {}, {
              group: v.props.label,
              children: (_v$props$children = v.props.children) === null || _v$props$children === void 0 ? void 0 : _v$props$children.map(function (v2) {
                return _handlerOptionElement(v2);
              })
            });
          }
          return _objectSpread(_objectSpread({}, v.props), {}, {
            label: v.props.label || v.props.children
          });
        }
        return {
          label: v
        };
      };
      transformedOptions = arrayChildren === null || arrayChildren === void 0 ? void 0 : arrayChildren.map(function (v) {
        return _handlerOptionElement(v);
      });
    }
    if (keys) {
      var _transformedOptions;
      transformedOptions = (_transformedOptions = transformedOptions) === null || _transformedOptions === void 0 ? void 0 : _transformedOptions.map(function (option) {
        return _objectSpread(_objectSpread({}, option), {}, {
          value: get(option, (keys === null || keys === void 0 ? void 0 : keys.value) || "value"),
          label: get(option, (keys === null || keys === void 0 ? void 0 : keys.label) || "label")
        });
      });
    }
    setCurrentOptions(transformedOptions);
    setTmpPropOptions(transformedOptions);
    setValueToOption(getValueToOption(children, options, keys) || {});
  }, [options, keys, children, reserveKeyword]);
  useEffect(function () {
    setSelectedOptions(function (oldSelectedOptions) {
      var valueKey = (keys === null || keys === void 0 ? void 0 : keys.value) || "value";
      var labelKey = (keys === null || keys === void 0 ? void 0 : keys.label) || "label";
      if (Array.isArray(value)) {
        return value.map(function (item) {
          if (valueType === "value") {
            return valueToOption[item] || oldSelectedOptions.find(function (option) {
              return get(option, valueKey) === item;
            }) || _defineProperty(_defineProperty({}, valueKey, item), labelKey, item);
          }
          return item;
        }).filter(Boolean);
      }
      if (value !== void 0 && value !== null) {
        if (valueType === "value") {
          return [valueToOption[value] || oldSelectedOptions.find(function (option) {
            return get(option, valueKey) === value;
          }) || _defineProperty(_defineProperty({}, valueKey, value), labelKey, value)].filter(Boolean);
        }
        return [value];
      }
      return [];
    });
  }, [value, keys, valueType, valueToOption, setSelectedOptions]);
  return {
    currentOptions: currentOptions,
    setCurrentOptions: setCurrentOptions,
    tmpPropOptions: tmpPropOptions,
    setTmpPropOptions: setTmpPropOptions,
    valueToOption: valueToOption,
    setValueToOption: setValueToOption,
    selectedOptions: selectedOptions,
    setSelectedOptions: setSelectedOptions
  };
}

export { UseOptions as default };
