// tdesign-react/cjs/comment/Comment.js
/**
 * tdesign v1.12.0
 * (c) 2025 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var classNames = require('classnames');
var button_index = require('../button/index.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var comment_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../button/Button.js');
require('../_chunks/dep-5481698f.js');
require('../_chunks/dep-dab07be2.js');
require('../_chunks/dep-b20af383.js');
require('../_chunks/dep-7225af62.js');
require('../_chunks/dep-d19a672c.js');
require('../hooks/useDomRefCallback.js');
require('../hooks/useRipple.js');
require('../hooks/useAnimation.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-abe9dd70.js');
require('../_chunks/dep-75a87f7c.js');
require('dayjs');
require('../_chunks/dep-f0496dae.js');
require('../_chunks/dep-89f1315e.js');
require('../_chunks/dep-49b692b4.js');
require('../_chunks/dep-5dc79997.js');
require('../_chunks/dep-350e0ec1.js');
require('../_chunks/dep-cef99009.js');
require('../_chunks/dep-a241cf5a.js');
require('../_chunks/dep-7a32c861.js');
require('../_chunks/dep-d4fd8ae6.js');
require('../_chunks/dep-de258415.js');
require('../_chunks/dep-b20ed9a4.js');
require('../_chunks/dep-ea572de0.js');
require('../_chunks/dep-ac56ab8f.js');
require('../_chunks/dep-e3f45bee.js');
require('../_chunks/dep-7ffb0d81.js');
require('../_chunks/dep-d02e27a7.js');
require('../_chunks/dep-dd8855d3.js');
require('../_chunks/dep-c2bb7040.js');
require('../_chunks/dep-ffef71d0.js');
require('../_chunks/dep-53b02ac8.js');
require('../_chunks/dep-54d84720.js');
require('../_chunks/dep-da766de6.js');
require('../_util/setStyle.js');
require('../_util/dom.js');
require('raf');
require('../_util/easing.js');
require('../_chunks/dep-b2af1b16.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../_chunks/dep-05ad44c3.js');
require('../_chunks/dep-be7119b7.js');
require('../_chunks/dep-b691f3a6.js');
require('../_chunks/dep-ca323c2b.js');
require('../loading/defaultProps.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-2b39f9ea.js');
require('react-dom/client');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-040a2db4.js');
require('../_chunks/dep-641538cb.js');
require('../_chunks/dep-b686b3fb.js');
require('../_chunks/dep-307d0db8.js');
require('../_chunks/dep-b3648f6a.js');
require('../_chunks/dep-62fd65d7.js');
require('../_chunks/dep-62978241.js');
require('../button/defaultProps.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-e831f387.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var Comment = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, comment_defaultProps.commentDefaultProps),
    actions = _useDefaultProps.actions,
    author = _useDefaultProps.author,
    avatar = _useDefaultProps.avatar,
    content = _useDefaultProps.content,
    datetime = _useDefaultProps.datetime,
    reply = _useDefaultProps.reply,
    quote = _useDefaultProps.quote,
    className = _useDefaultProps.className,
    style = _useDefaultProps.style;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var avatarElement = avatar ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__avatar")
  }, typeof avatar === "string" ? /* @__PURE__ */React__default["default"].createElement("img", {
    src: avatar,
    alt: "",
    className: "".concat(classPrefix, "-comment__avatar-image")
  }) : avatar) : null;
  var authorDatetimeContent = (author || datetime) && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__author")
  }, author && /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-comment__name")
  }, author), datetime && /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-comment__time")
  }, datetime));
  var quoteElement = quote ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__quote")
  }, quote) : null;
  var actionsElement = actions && actions.length ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__actions")
  }, actions.map(function (action, index) {
    return /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
      key: "action-".concat(index),
      size: "small",
      variant: "text"
    }, action);
  })) : null;
  var contentElement = /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__content")
  }, authorDatetimeContent, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__detail")
  }, content), quoteElement, actionsElement);
  var replyElement = reply ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-comment__reply"))
  }, reply) : null;
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    style: style,
    className: classNames__default["default"]("".concat(classPrefix, "-comment"), className)
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__inner")
  }, avatarElement, contentElement), replyElement);
});
Comment.displayName = "Comment";

exports["default"] = Comment;
